---
title: "7_figure6_plots"
output: rmarkdown::github_document
date: '2022-07-10'
---

The purpose of this figure is to show that circulating CD4+ and CD8+ T cells in the blood that have matching TCRs to cells in the pancreas show unique transcriptional features in checkpoint-induced T1D compared to spontaneous T1D

```{r setup, include=FALSE, echo=FALSE}
knitr::opts_chunk$set(echo = TRUE, root.dir = "C:/Users/Strix/GitRepos/NOD_spt_pd1")
knitr::opts_knit$set(echo = TRUE, root.dir = "C:/Users/Strix/GitRepos/NOD_spt_pd1")
```

```{r libraries, message = FALSE, warning = FALSE}
require(dplyr, quietly = T) #simple manipulation of data.frames
require(Seurat, quietly = T) #scRNA-seq analysis
require(reshape, quietly = T) #melting and casting data.frames
require(ggplot2, quietly = T) #plotting
require(rstatix, quietly = T) #adding p values to plots
require(ggprism, quietly = T) #adding p values to plots
require(ggsci, quirtly = T) #IGV color palette
require(patchwork, quietly = T) #adding p values to plots
require(magrittr, quietly = T) #adding p values to plots
require(gridExtra, quietly = T) #display multiple plots
require(ggrepel, quietly = T) #labels for plots
require(tibble, quietly = T) #rownames_to_column function
require(useful, quietly = T) #corner function to peek at data.frames
require(viridis, quietly = T) #viridis color palettes
require(readxl, quietly = T) #reading Excel files
require(msigdbr, quietly = T) #accessing gene signatures from mSigDB
source("scripts/utils.R") #custom color palettes and other functions
set.seed(1) #set a seed for reproducibility
```

```{r read in Seurats}
seurat_integrated <- saveRDS("objects/seurat_integrated_clustered.rds")
pan4 <- readRDS("objects/seurat_cd4_panonly_clustered.rds")
pan8 <- readRDS("objects/seurat_cd8_panonly_clustered.rds")
```

First, UMAPs showing the distribution of PM and non-PM T cells in the blood will be plotted, followed by some pie charts illustrating the distribution of clusters for each group.
```{r UMAPs of shared/non-shared blood}
p_blood_shared <- seurat_integrated@meta.data %>% 
                    filter(tissue == "blood" & !is.na(pancreas_matching)) %>%
                    ggplot(aes(x = allcells_UMAP_1, 
                               y = allcells_UMAP_2, 
                               color = seurat_clusters)) + 
                    geom_point(data = seurat_integrated@meta.data %>% 
                                 dplyr::select(-pancreas_matching), 
                               shape = 16, 
                               color = "grey90") +
                    geom_point(shape = 16, 
                               alpha = 0.3) +
                    theme_minimal() +
                    theme(axis.title = element_blank(),
                          axis.text = element_blank(),
                          panel.grid = element_blank(),
                          text = element_text(size = 15),
                          legend.position = "none") +
                    labs(color=NULL,
                         title = "T cells in blood: pancreas-matching (PM) or non-matching") +
                    scale_color_manual(values = custom_colors("clusters")) +
                    guides(color = guide_legend(override.aes = list(alpha = 1, 
                                                                  size = 5, 
                                                                  shape = 19))) +
                  facet_grid(rows = vars(pancreas_matching), 
                             cols = vars(treatment))
p_blood_shared

#PIES OF CLUSTERS IN EACH GROUP
p_cd8_blood_pie <- seurat_integrated@meta.data %>% 
                    filter(tissue == "blood" & !is.na(pancreas_matching) & celltype == "CD8") %>%
                    group_by(treatment,pancreas_matching,
                             seurat_clusters) %>% 
                    tally() %>% 
                    group_by(treatment, pancreas_matching) %>% 
                    mutate(prcnt = n*100/sum(n),
                           csum = rev(cumsum(rev(prcnt))),  #get the position for the labels
                           pos = prcnt/2 + lead(csum, 1),
                           pos = if_else(is.na(pos), prcnt/2, pos)) %>%
                    ggplot(aes(x = "", 
                               y = prcnt, 
                               fill = seurat_clusters)) + # pct used here so slices add to 100
                    geom_bar(stat="identity", width=1) +
                    coord_polar("y", 
                                start = 0) +
                    #geom_text(aes(label = paste0(round(prcnt, 0), "%")), position = position_stack(vjust = 0.5)) +
                    facet_grid(rows = vars(pancreas_matching), 
                               cols = vars(treatment)) +
                    theme_void() +
                    theme(legend.position = "none",
                          text = element_text(size = 15)) +
                    labs(title = "CD8+ T cells in blood") +
                    geom_label_repel(aes(y = pos, 
                                         label = paste0(round(prcnt, 0), "%")),
                                         size = 4.5, 
                                         nudge_x = 0.1, 
                                         show.legend = FALSE) +
                    scale_fill_manual(values = custom_colors("clusters"))
p_cd4_blood_pie <- seurat_integrated@meta.data %>% 
                    filter(tissue == "blood" & !is.na(pancreas_matching) & celltype %in% c("Tcon","Treg")) %>%
                    group_by(treatment,pancreas_matching,
                             seurat_clusters) %>% 
                    tally() %>% 
                    group_by(treatment, pancreas_matching) %>% 
                    mutate(prcnt = n*100/sum(n),
                           csum = rev(cumsum(rev(prcnt))),  #get the position for the labels
                           pos = prcnt/2 + lead(csum, 1),
                           pos = if_else(is.na(pos), prcnt/2, pos)) %>%
                    ggplot(aes(x = "", 
                               y = prcnt, 
                               fill = seurat_clusters)) + # pct used here so slices add to 100
                    geom_bar(stat="identity", width=1) +
                    coord_polar("y", 
                                start = 0) +
                    #geom_text(aes(label = paste0(round(prcnt, 0), "%")), position = position_stack(vjust = 0.5)) +
                    facet_grid(rows = vars(pancreas_matching), 
                               cols = vars(treatment)) +
                    theme_void() +
                    theme(legend.position = "none",
                          text = element_text(size = 15)) +
                    labs(title = "CD4+ T cells in blood") +
                    geom_label_repel(aes(y = pos, 
                                         label = paste0(round(prcnt, 0), "%")),
                                         size = 4.5, 
                                         nudge_x = 0.1, 
                                         show.legend = FALSE) +
                    scale_fill_manual(values = custom_colors("clusters"))
grid.arrange(p_cd8_blood_pie, p_cd4_blood_pie, nrow = 2)
```

Show the top 50 clonotypes as a percentage of TCRs detected in the PANCREAS as a cumulative sum and also the clone sizes for the different treatment groups.
```{r top 50 clonotypes and clone size in pancreas}
#TOP 50 CLONOTYPES: CD8 IN PANCREAS
cumulative_sums <- seurat_integrated@meta.data %>% 
                    filter(clonotype_tissue_groups %in% c("blood&pancreas","pancreas") & celltype == "CD8") %>%
                      group_by(treatment, clonotype_id, pancreas_matching) %>% 
                      tally() %>% 
                      group_by(treatment) %>% 
                      mutate(prcnt = n*100/sum(n)) %>% 
                      group_by(treatment, pancreas_matching) %>% 
                      arrange(treatment, pancreas_matching, desc(prcnt)) %>% 
                      mutate(cs = cumsum(prcnt))

cumulative_sums$clonotype_id <- factor(cumulative_sums$clonotype_id, 
                                       levels = cumulative_sums$clonotype_id) #orders the clonotypes on x axis

p_cd8_panPM_top50_clonos <- cumulative_sums %>% 
                              group_by(treatment, pancreas_matching) %>% 
                              slice_head(n = 50) %>% 
                              ggplot(aes(x = clonotype_id, 
                                         y = cs, 
                                         color = treatment)) + 
                              geom_point() + 
                              geom_segment(aes(xend = clonotype_id, 
                                               yend = cs, 
                                               x = clonotype_id, 
                                               y = 0, 
                                               color = treatment)) + 
                              scale_color_manual(values = custom_colors("treatments")) +
                              scale_y_continuous(expand = c(0, 0)) +
                              theme_minimal() +
                              theme(panel.grid = element_blank(),
                                    panel.spacing.x = unit(0, "lines"),
                                    text = element_text(size = 15),
                                    axis.text.x = element_blank(),
                                    panel.border = element_rect(colour = "black", 
                                                                fill = NA, 
                                                                size = 1),
                                    legend.position = "none") +
                              labs(x = "Top 50 clonotypes", 
                                   y = "Cumulative % CD8 TCRs",
                                   title = "CD8+ T cells in pancreas") +
                              facet_wrap(treatment ~ pancreas_matching, scales = "free_x", ncol = 2)

#TOP 50 CLONOTYPES: TCON IN PANCREAS
cumulative_sums <- seurat_integrated@meta.data %>% 
                    filter(clonotype_tissue_groups %in% c("blood&pancreas","pancreas") & celltype == "Tcon") %>%
                      group_by(treatment, clonotype_id, pancreas_matching) %>% 
                      tally() %>% 
                      group_by(treatment) %>% 
                      mutate(prcnt = n*100/sum(n)) %>% 
                      group_by(treatment, pancreas_matching) %>% 
                      arrange(treatment, pancreas_matching, desc(prcnt)) %>% 
                      mutate(cs = cumsum(prcnt))

cumulative_sums$clonotype_id <- factor(cumulative_sums$clonotype_id, 
                                       levels = cumulative_sums$clonotype_id) #orders the clonotypes on x axis

p_tcon_panPM_top50_clonos <- cumulative_sums %>% 
                              group_by(treatment, pancreas_matching) %>% 
                              slice_head(n = 50) %>% 
                              ggplot(aes(x = clonotype_id, 
                                         y = cs, 
                                         color = treatment)) + 
                              geom_point() + 
                              geom_segment(aes(xend = clonotype_id, 
                                               yend = cs, 
                                               x = clonotype_id, 
                                               y = 0, 
                                               color = treatment)) + 
                              scale_color_manual(values = custom_colors("treatments")) +
                              scale_y_continuous(expand = c(0, 0)) +
                              theme_minimal() +
                              theme(panel.grid = element_blank(),
                                    panel.spacing.x = unit(0, "lines"),
                                    text = element_text(size = 15),
                                    axis.text.x = element_blank(),
                                    panel.border = element_rect(colour = "black", 
                                                                fill = NA, 
                                                                size = 1),
                                    legend.position = "none") +
                              labs(x = "Top 50 clonotypes", 
                                   y = "Cumulative % Tcon TCRs",
                                   title = "Tcon in pancreas") +
                              facet_wrap(treatment ~ pancreas_matching, scales = "free_x", ncol = 2)

#TOP 50 CLONOTYPES: TREG IN PANCREAS
cumulative_sums <- seurat_integrated@meta.data %>% 
                    filter(clonotype_tissue_groups %in% c("blood&pancreas","pancreas") & celltype == "Treg") %>%
                      group_by(treatment, clonotype_id, pancreas_matching) %>% 
                      tally() %>% 
                      group_by(treatment) %>% 
                      mutate(prcnt = n*100/sum(n)) %>% 
                      group_by(treatment, pancreas_matching) %>% 
                      arrange(treatment, pancreas_matching, desc(prcnt)) %>% 
                      mutate(cs = cumsum(prcnt))

cumulative_sums$clonotype_id <- factor(cumulative_sums$clonotype_id, 
                                       levels = cumulative_sums$clonotype_id) #orders the clonotypes on x axis

p_treg_panPM_top50_clonos <- cumulative_sums %>% 
                              group_by(treatment, pancreas_matching) %>% 
                              slice_head(n = 50) %>% 
                              ggplot(aes(x = clonotype_id, 
                                         y = cs, 
                                         color = treatment)) + 
                              geom_point() + 
                              geom_segment(aes(xend = clonotype_id, 
                                               yend = cs, 
                                               x = clonotype_id, 
                                               y = 0, 
                                               color = treatment)) + 
                              scale_color_manual(values = custom_colors("treatments")) +
                              scale_y_continuous(expand = c(0, 0)) +
                              theme_minimal() +
                              theme(panel.grid = element_blank(),
                                    panel.spacing.x = unit(0, "lines"),
                                    text = element_text(size = 15),
                                    axis.text.x = element_blank(),
                                    panel.border = element_rect(colour = "black", 
                                                                fill = NA, 
                                                                size = 1),
                                    legend.position = "none") +
                              labs(x = "Top 50 clonotypes", 
                                   y = "Cumulative % Treg TCRs",
                                   title = "Treg in pancreas") +
                              facet_wrap(treatment ~ pancreas_matching, scales = "free_x", ncol = 2)

grid.arrange(p_cd8_panPM_top50_clonos, p_tcon_panPM_top50_clonos, p_treg_panPM_top50_clonos, nrow = 1)

#BOXPLOT OF CLONE SIZE
stat.test <- data.frame(seurat_integrated@meta.data %>%
                          filter(tissue == "pancreas" & !is.na(pancreas_matching)) %>%
                          group_by(treatment, pancreas_matching, clonotype_id) %>%
                          tally() %>%
                          group_by(treatment) %>%
                          wilcox_test(n ~ pancreas_matching) %>% 
                          add_xy_position() %>%
                          mutate(y.position = log10(y.position + 1)))

p_boxplot_pm_pan <- seurat_integrated@meta.data %>%
    filter(tissue == "pancreas" & !is.na(pancreas_matching)) %>%
    group_by(celltype, treatment, pancreas_matching, clonotype_id) %>%
    tally() %>%
    ggplot() + 
    geom_boxplot(aes(x = pancreas_matching, 
               y = log10(n), 
               fill = treatment,
               color = treatment),
               outlier.shape = NA, 
               alpha = 0.5,
               position = position_dodge(0.9)) +
    geom_point(aes(x = pancreas_matching, 
               y = log10(n), 
               fill = treatment,
               color = treatment),
               position = position_jitterdodge(dodge.width = 0.9,
                                               jitter.width = 0.8),
               alpha = 0.1) +
    scale_fill_manual(values = custom_colors("treatments")) +
    scale_color_manual(values = custom_colors("treatments")) +
    theme_minimal() +
    theme(panel.grid = element_blank(),
          axis.title.x = element_blank(),
          text = element_text(size = 15),
          legend.position = "top",
          panel.border = element_rect(colour = "black", fill = NA, size = 1)) +
    labs(y = expression(log[10](clone~size)),
         color = NULL,
         fill = NULL,
         title = "T cells in pancreas (unique clonotypes)") +
    facet_wrap(~celltype, nrow = 1)

#DIVERSITY ANALYSIS
# tcrtypecounts <- rbind(data.frame(pan8@meta.data %>% dplyr::select(treatment, mouse_id, clonotype_id, clonotype_count)))
# corner(tcrtypecounts)
# 
# tcrtypecounts <- tcrtypecounts %>% filter(!is.na(clonotype_id)) %>% cast(treatment + mouse_id  ~ clonotype_id)
# corner(tcrtypecounts)
# 
# totalN <- tcrtypecounts %>% select(-c(treatment, mouse_id)) %>% rowSums() # Total N by row sums
# df <- sapply(tcrtypecounts %>% select(-c(treatment, mouse_id)), function(x) (x - 1)*x) # Calculate x(x-1) for each value in df
# corner(df)
# numerator <- df %>% rowSums() # Calculate numerator of equation (x in above)
# 
# # Calculate simpson index
# diversities <- data.frame(tcrtypecounts %>% 
#                             dplyr::select(c(treatment, mouse_id)), 
#                           celltype = "CD8", 
#                           simpson = totalN*(totalN - 1) / numerator)
# 
# diversities$treatment <- factor(diversities$treatment, levels = c("IgG", "Spt", "PD1")) #factor the pathologies
# 
# stat.test <- data.frame(diversities %>% 
#                             mutate(simpson = ifelse(simpson == Inf, NA, simpson)) %>% 
#                             wilcox_test(simpson ~ treatment) %>% 
#                             add_xy_position())
# 
# diversities %>%
#     ggplot(aes(x = treatment, 
#                y = simpson,
#                fill = treatment, 
#                color = treatment)) + 
#     geom_violin(alpha = 0.5,
#                 adjust = 1,
#                 scale = "width",
#                 trim = F) + 
#     geom_point(size = 5) +
#     labs(y = "Simpson index") +
#     theme_minimal() +
#     theme(panel.grid = element_blank(),
#           text = element_text(size = 15),
#           panel.border = element_rect(colour = "black", 
#                                       fill=NA, 
#                                       size=1),
#           legend.position = "bottom") +
#     labs(title = "CD8 T cells in pancreas") +
#     scale_fill_manual(values = custom_colors("treatments")) +
#     scale_color_manual(values = custom_colors("treatments")) +
#     add_pvalue(stat.test,
#                inherit.aes = F,
#                label = "p.adj.signif",
#                 y.position = "y.position") 

```

Show the top 50 clonotypes as a percentage of TCRs detected in the BLOOD as a cumulative sum and also the clone sizes for the different treatment groups.
```{r top 50 clonotypes and clone size in blood}
#TOP 50 CLONOTYPES: CD8 IN BLOOD
cumulative_sums <- seurat_integrated@meta.data %>% 
                    filter(clonotype_tissue_groups %in% c("blood&pancreas","blood") & celltype == "CD8") %>%
                      group_by(treatment, clonotype_id, pancreas_matching) %>% 
                      tally() %>% 
                      group_by(treatment) %>% 
                      mutate(prcnt = n*100/sum(n)) %>% 
                      group_by(treatment, pancreas_matching) %>% 
                      arrange(treatment, pancreas_matching, desc(prcnt)) %>% 
                      mutate(cs = cumsum(prcnt))

cumulative_sums$clonotype_id <- factor(cumulative_sums$clonotype_id, 
                                       levels = cumulative_sums$clonotype_id) #orders the clonotypes on x axis

p_cd8_bldPM_top50_clonos <- cumulative_sums %>% 
                              group_by(treatment, pancreas_matching) %>% 
                              slice_head(n = 50) %>% 
                              ggplot(aes(x = clonotype_id, 
                                         y = cs, 
                                         color = treatment)) + 
                              geom_point() + 
                              geom_segment(aes(xend = clonotype_id, 
                                               yend = cs, 
                                               x = clonotype_id, 
                                               y = 0, 
                                               color = treatment)) + 
                              scale_color_manual(values = custom_colors("treatments")) +
                              scale_y_continuous(expand = c(0, 0)) +
                              theme_minimal() +
                              theme(panel.grid = element_blank(),
                                    panel.spacing.x = unit(0, "lines"),
                                    text = element_text(size = 15),
                                    axis.text.x = element_blank(),
                                    panel.border = element_rect(colour = "black", 
                                                                fill = NA, 
                                                                size = 1),
                                    legend.position = "none") +
                              labs(x = "Top 50 clonotypes", 
                                   y = "Cumulative % CD8 TCRs",
                                   title = "CD8+ T cells in blood") +
                              facet_wrap(treatment ~ pancreas_matching, scales = "free_x", ncol = 2)

#TOP 50 CLONOTYPES: TCON IN BLOOD
cumulative_sums <- seurat_integrated@meta.data %>% 
                    filter(clonotype_tissue_groups %in% c("blood&pancreas","blood") & celltype == "Tcon") %>%
                      group_by(treatment, clonotype_id, pancreas_matching) %>% 
                      tally() %>% 
                      group_by(treatment) %>% 
                      mutate(prcnt = n*100/sum(n)) %>% 
                      group_by(treatment, pancreas_matching) %>% 
                      arrange(treatment, pancreas_matching, desc(prcnt)) %>% 
                      mutate(cs = cumsum(prcnt))

cumulative_sums$clonotype_id <- factor(cumulative_sums$clonotype_id, 
                                       levels = cumulative_sums$clonotype_id) #orders the clonotypes on x axis

p_tcon_bldPM_top50_clonos <- cumulative_sums %>% 
                              group_by(treatment, pancreas_matching) %>% 
                              slice_head(n = 50) %>% 
                              ggplot(aes(x = clonotype_id, 
                                         y = cs, 
                                         color = treatment)) + 
                              geom_point() + 
                              geom_segment(aes(xend = clonotype_id, 
                                               yend = cs, 
                                               x = clonotype_id, 
                                               y = 0, 
                                               color = treatment)) + 
                              scale_color_manual(values = custom_colors("treatments")) +
                              scale_y_continuous(expand = c(0, 0)) +
                              theme_minimal() +
                              theme(panel.grid = element_blank(),
                                    panel.spacing.x = unit(0, "lines"),
                                    text = element_text(size = 15),
                                    axis.text.x = element_blank(),
                                    panel.border = element_rect(colour = "black", 
                                                                fill = NA, 
                                                                size = 1),
                                    legend.position = "none") +
                              labs(x = "Top 50 clonotypes", 
                                   y = "Cumulative % Tcon TCRs",
                                   title = "Tcon in blood") +
                              facet_wrap(treatment ~ pancreas_matching, scales = "free_x", ncol = 2)

#TOP 50 CLONOTYPES: TREG IN PANCREAS
cumulative_sums <- seurat_integrated@meta.data %>% 
                    filter(clonotype_tissue_groups %in% c("blood&pancreas","blood") & celltype == "Treg") %>%
                      group_by(treatment, clonotype_id, pancreas_matching) %>% 
                      tally() %>% 
                      group_by(treatment) %>% 
                      mutate(prcnt = n*100/sum(n)) %>% 
                      group_by(treatment, pancreas_matching) %>% 
                      arrange(treatment, pancreas_matching, desc(prcnt)) %>% 
                      mutate(cs = cumsum(prcnt))

cumulative_sums$clonotype_id <- factor(cumulative_sums$clonotype_id, 
                                       levels = cumulative_sums$clonotype_id) #orders the clonotypes on x axis

p_treg_bldPM_top50_clonos <- cumulative_sums %>% 
                              group_by(treatment, pancreas_matching) %>% 
                              slice_head(n = 50) %>% 
                              ggplot(aes(x = clonotype_id, 
                                         y = cs, 
                                         color = treatment)) + 
                              geom_point() + 
                              geom_segment(aes(xend = clonotype_id, 
                                               yend = cs, 
                                               x = clonotype_id, 
                                               y = 0, 
                                               color = treatment)) + 
                              scale_color_manual(values = custom_colors("treatments")) +
                              scale_y_continuous(expand = c(0, 0)) +
                              theme_minimal() +
                              theme(panel.grid = element_blank(),
                                    panel.spacing.x = unit(0, "lines"),
                                    text = element_text(size = 15),
                                    axis.text.x = element_blank(),
                                    panel.border = element_rect(colour = "black", 
                                                                fill = NA, 
                                                                size = 1),
                                    legend.position = "none") +
                              labs(x = "Top 50 clonotypes", 
                                   y = "Cumulative % Treg TCRs",
                                   title = "Treg in blood") +
                              facet_wrap(treatment ~ pancreas_matching, scales = "free_x", ncol = 2)

grid.arrange(p_cd8_bldPM_top50_clonos, p_tcon_bldPM_top50_clonos, p_treg_bldPM_top50_clonos, nrow = 1)

#BOXPLOT OF CLONE SIZE
stat.test <- data.frame(seurat_integrated@meta.data %>%
                          filter(tissue == "blood" & !is.na(pancreas_matching)) %>%
                          group_by(treatment, pancreas_matching, clonotype_id) %>%
                          tally() %>%
                          group_by(treatment) %>%
                          wilcox_test(n ~ pancreas_matching) %>% 
                          add_xy_position() %>%
                          mutate(y.position = log10(y.position + 1)))

p_boxplot_pm_bld <- seurat_integrated@meta.data %>%
    filter(tissue == "blood" & !is.na(pancreas_matching)) %>%
    group_by(celltype, treatment, pancreas_matching, clonotype_id) %>%
    tally() %>% 
    ggplot() + 
    geom_boxplot(aes(x = pancreas_matching, 
               y = log10(n), 
               fill = treatment,
               color = treatment),
               outlier.shape = NA, 
               alpha = 0.5,
               position = position_dodge(0.9)) +
    geom_point(aes(x = pancreas_matching, 
               y = log10(n), 
               fill = treatment,
               color = treatment),
               position = position_jitterdodge(dodge.width = 0.9,
                                               jitter.width = 0.8),
               alpha = 0.1) +
    scale_fill_manual(values = custom_colors("treatments")) +
    scale_color_manual(values = custom_colors("treatments")) +
    theme_minimal() +
    theme(panel.grid = element_blank(),
          axis.title.x = element_blank(),
          text = element_text(size = 15),
          legend.position = "top",
          panel.border = element_rect(colour = "black", fill = NA, size = 1)) +
    labs(y = expression(log[10](clone~size)),
         color = NULL,
         fill = NULL,
         title = "T cells in blood (unique clonotypes)") +
    facet_wrap(~celltype, nrow = 1)

#DIVERSITY ANALYSIS
# tcrtypecounts <- rbind(data.frame(pan8@meta.data %>% dplyr::select(treatment, mouse_id, clonotype_id, clonotype_count)))
# corner(tcrtypecounts)
# 
# tcrtypecounts <- tcrtypecounts %>% filter(!is.na(clonotype_id)) %>% cast(treatment + mouse_id  ~ clonotype_id)
# corner(tcrtypecounts)
# 
# totalN <- tcrtypecounts %>% select(-c(treatment, mouse_id)) %>% rowSums() # Total N by row sums
# df <- sapply(tcrtypecounts %>% select(-c(treatment, mouse_id)), function(x) (x - 1)*x) # Calculate x(x-1) for each value in df
# corner(df)
# numerator <- df %>% rowSums() # Calculate numerator of equation (x in above)
# 
# # Calculate simpson index
# diversities <- data.frame(tcrtypecounts %>% 
#                             dplyr::select(c(treatment, mouse_id)), 
#                           celltype = "CD8", 
#                           simpson = totalN*(totalN - 1) / numerator)
# 
# diversities$treatment <- factor(diversities$treatment, levels = c("IgG", "Spt", "PD1")) #factor the pathologies
# 
# stat.test <- data.frame(diversities %>% 
#                             mutate(simpson = ifelse(simpson == Inf, NA, simpson)) %>% 
#                             wilcox_test(simpson ~ treatment) %>% 
#                             add_xy_position())
# 
# diversities %>%
#     ggplot(aes(x = treatment, 
#                y = simpson,
#                fill = treatment, 
#                color = treatment)) + 
#     geom_violin(alpha = 0.5,
#                 adjust = 1,
#                 scale = "width",
#                 trim = F) + 
#     geom_point(size = 5) +
#     labs(y = "Simpson index") +
#     theme_minimal() +
#     theme(panel.grid = element_blank(),
#           text = element_text(size = 15),
#           panel.border = element_rect(colour = "black", 
#                                       fill=NA, 
#                                       size=1),
#           legend.position = "bottom") +
#     labs(title = "CD8 T cells in pancreas") +
#     scale_fill_manual(values = custom_colors("treatments")) +
#     scale_color_manual(values = custom_colors("treatments")) +
#     add_pvalue(stat.test,
#                inherit.aes = F,
#                label = "p.adj.signif",
#                 y.position = "y.position") 

```

Plot the correlation of clonotype sizes in both pancreas and blood
```{r correlation of clonotype size in pancreas and blood}
seurat_integrated@meta.data %>% 
    filter(clonotype_tissue_groups %in% c("blood&pancreas","pLN&blood&pancreas") & tissue %in% c("blood","pancreas")) %>%
    group_by(treatment, cd4_or_cd8, clonotype_id, tissue) %>% 
    tally() %>% 
    cast(treatment+cd4_or_cd8+clonotype_id~tissue, value = "n") %>%
    group_by(treatment, cd4_or_cd8, blood, pancreas) %>%
    tally(name = "number_of_clonotypes") %>%
    ggplot(aes(x = log10(pancreas),
               y = log10(blood),
               size = number_of_clonotypes,
               color = treatment)) +
    geom_point() +
    theme_minimal() +
    facet_wrap(cd4_or_cd8~treatment) + 
    scale_color_manual(values = custom_colors("treatments")) +
    theme_minimal() +
    theme(panel.grid = element_blank(),
          text = element_text(size = 15),
          panel.spacing.x = unit(0, "lines"),
          legend.position = "right",
          panel.border = element_rect(colour = "black", fill = NA, size = 1)) +
    labs(y = expression(log[10](blood~clone~size)),
         x = expression(log[10](pancreas~clone~size)),
         color = "Treatment",
         title = "T cells in blood",
         size = "Number of\nunique clonotypes") +
  scale_size_continuous(range = c(2,10))
```

DE genes volcano plot within pancreas matching cells in blood:
```{r volcano plot of PM T cells in blood}
cutoff <- 0.25 #set a cutoff for DE gene expression
ident1 <- "PD1"
ident2 <- "Spt"

Idents(seurat_integrated) <- seurat_integrated$tissue
seurat_bld <- subset(seurat_integrated, idents = "blood") #subset on T cells in blood
Idents(seurat_bld) <- seurat_bld$pancreas_matching
seurat_bld <- subset(seurat_bld, idents = "PM") #subset on PM T cells only
Idents(seurat_bld) <- seurat_bld$celltype

volcanodata <- list()
allvolcano <- data.frame()
for (current_cell in c("CD8", "Tcon", "Treg")) {
  volcanodata[[current_cell]] <- FindMarkers(seurat_bld, #define the table of DE genes
                                     ident.1 = ident1,
                                     ident.2 = ident2, 
                                     verbose = F,
                                     group.by = "treatment",
                                     subset.ident = current_cell,
                                     only.pos = F, 
                                     logfc.threshold = cutoff, 
                                     assay = "RNA") %>% 
                                  rownames_to_column("gene") %>%
                                  mutate(celltype = current_cell)
  #assess significance
  volcanodata[[current_cell]]$DE <- "None"
  volcanodata[[current_cell]][which(volcanodata[[current_cell]]$p_val_adj < 0.05 & 
                                      volcanodata[[current_cell]]$avg_log2FC >= cutoff), "DE"] <- ident1 #positive avg_log2FC are associated with ident.1
  volcanodata[[current_cell]][which(volcanodata[[current_cell]]$p_val_adj < 0.05 & 
                                      volcanodata[[current_cell]]$avg_log2FC <= -cutoff), "DE"] <- ident2 #negative avg_log2FC are associated with ident.2
  allvolcano <- rbind(allvolcano, volcanodata[[current_cell]])
}

#label individual genes
delabels <- allvolcano %>% filter(DE != 'None') #only label significant genes
delabels <- delabels %>% filter(gene %in% goi) #filter based on genes of interest

allvolcano %>%
    ggplot(aes(x = avg_log2FC,
               y = -log10(p_val_adj))) + 
        geom_hline(yintercept = -log10(0.05),
                   size = 0.5,
                   color = "grey90",
                   alpha = 0.5) + 
        geom_vline(xintercept = c(-cutoff, cutoff),
                   size = 0.5,
                   color = "grey90",
                   alpha = 0.5) + 
        geom_point(aes(color = DE),
                   size = 3,
                   alpha = 0.5,
                   shape = 16) +
        scale_color_manual(values = c("None" = "grey70",
                                      custom_colors("treatments"))) +
        scale_y_continuous(limits = c(0,10)) +
        theme_minimal() +
        theme(panel.grid = element_blank(),
              text = element_text(size = 15),
              panel.border = element_rect(fill = NA, color = "black"),
              axis.text = element_text(), 
              panel.spacing = unit(0, "lines"),
              axis.title = element_text()) +
        labs(x = paste("Average Log2 Fold Change (",ident1,"/",ident2,")"),
             y = expression(-log[10]~(Adjusted~p~Value)),
             title="Pancreas-matching T cells in blood") +
        geom_label_repel(data = delabels,
                         aes(x = avg_log2FC,
                             y = -log10(p_val_adj),
                             label = gene,
                             color = DE),
                         label.size = NA,
                         label.padding = 0.2, 
                         na.rm = TRUE,
                         fill = alpha(c("white"), 0.8), 
                         size=7, 
                         max.overlaps = 50) +
  facet_wrap(~celltype)
```

DE genes volcano plot between pancreas-matching and non-matching T cells in the blood between different treatment groups.
```{r volcano plot of PM vs. non-PM T cells in blood}
cutoff <- 0.25 #set a cutoff for DE gene expression
ident1 <- "PM"
ident2 <- "non-PM"

Idents(seurat_integrated) <- seurat_integrated$tissue
seurat_bld <- subset(seurat_integrated, idents = "blood") #subset on T cells in blood
Idents(seurat_bld) <- seurat_bld$treatment

allvolcano <- data.frame()
for (current_treatment in c("IgG","Spt","PD1")) {
  seurat_tmp <- subset(seurat_bld, idents = current_treatment) #subset on PM T cells only
  Idents(seurat_tmp) <- seurat_tmp$celltype
  
  volcanodata <- list()
  for (current_cell in c("CD8", "Tcon", "Treg")) {
    volcanodata[[current_cell]] <- FindMarkers(seurat_tmp, #define the table of DE genes
                                       ident.1 = ident1,
                                       ident.2 = ident2, 
                                       verbose = F,
                                       group.by = "pancreas_matching",
                                       subset.ident = current_cell,
                                       only.pos = F, 
                                       logfc.threshold = cutoff, 
                                       assay = "RNA") %>% 
                                    rownames_to_column("gene") %>%
                                    mutate(celltype = current_cell,
                                           treatment = current_treatment)
    #assess significance
    volcanodata[[current_cell]]$DE <- "None"
    volcanodata[[current_cell]][which(volcanodata[[current_cell]]$p_val_adj < 0.05 & 
                                        volcanodata[[current_cell]]$avg_log2FC >= cutoff), "DE"] <- ident1 #positive avg_log2FC are associated with ident.1
    volcanodata[[current_cell]][which(volcanodata[[current_cell]]$p_val_adj < 0.05 & 
                                        volcanodata[[current_cell]]$avg_log2FC <= -cutoff), "DE"] <- ident2 #negative avg_log2FC are associated with ident.2
    allvolcano <- rbind(allvolcano, volcanodata[[current_cell]])
  }
}
#label individual genes
delabels <- allvolcano %>% filter(DE != 'None') #only label significant genes
delabels <- delabels %>% filter(gene %in% goi) #filter based on genes of interest

allvolcano %>%
    ggplot(aes(x = avg_log2FC,
               y = -log10(p_val_adj))) + 
        geom_hline(yintercept = -log10(0.05),
                   size = 0.5,
                   color = "grey90",
                   alpha = 0.5) + 
        geom_vline(xintercept = c(-cutoff, cutoff),
                   size = 0.5,
                   color = "grey90",
                   alpha = 0.5) + 
        geom_point(aes(color = DE),
                   size = 3,
                   alpha = 0.5,
                   shape = 16) +
        scale_color_manual(values = c("None" = "grey70",
                                      custom_colors("pm"))) +
        scale_y_continuous() +
        theme_minimal() +
        theme(panel.grid = element_blank(),
              text = element_text(size = 15),
              panel.border = element_rect(fill = NA, color = "black"),
              axis.text = element_text(), 
              panel.spacing = unit(0, "lines"),
              axis.title = element_text()) +
        labs(x = paste("Average Log2 Fold Change (",ident1,"/",ident2,")"),
             y = expression(-log[10]~(Adjusted~p~Value)),
             title="T cells in blood") +
        geom_label_repel(data = delabels,
                         aes(x = avg_log2FC,
                             y = -log10(p_val_adj),
                             label = gene,
                             color = DE),
                         label.size = NA,
                         label.padding = 0.2, 
                         na.rm = TRUE,
                         fill = alpha(c("white"), 0.8), 
                         size = 5, 
                         max.overlaps = 50) +
  facet_wrap(celltype ~ treatment)
```

SAVE data for COMETSC
```{r COMET}
bld <- subset(seurat_integrated,idents = "blood")
Idents(bld) <- bld$cd4_or_cd8
bld8 <- subset(bld, idents = "CD8")
bld8 <- subset(bld8, cells = bld@meta.data %>% filter(!is.na(pancreas_matching)) %>% rownames())
write.table(bld8@meta.data %>% mutate(comet_clusters = paste0(treatment,"_",pancreas_matching))%>%dplyr::select(comet_clusters), "output/cometsc/cluster.txt", quote = F, sep = "\t", col.names = F)
write.table(bld8@meta.data %>% dplyr::select(allcells_UMAP_1,allcells_UMAP_2), "output/cometsc/vis.txt", quote = F, sep = "\t", col.names = F)
write.table(as.data.frame(bld8@assays$RNA@data), "output/cometsc/marker.txt", quote = F, sep = "\t")

```

Upset plot for sharing between tissues

