---
title: "7_figure6_plots"
output: rmarkdown::github_document
date: '2022-07-10'
---

The purpose of this figure is to show that circulating CD4+ and CD8+ T cells in the blood that have matching TCRs to cells in the pancreas show unique transcriptional features in checkpoint-induced T1D compared to spontaneous T1D

```{r setup, include=FALSE, echo=FALSE}
knitr::opts_chunk$set(echo = TRUE, root.dir = "C:/Users/Strix/GitRepos/NOD_spt_pd1")
knitr::opts_knit$set(echo = TRUE, root.dir = "C:/Users/Strix/GitRepos/NOD_spt_pd1")
```

```{r libraries, message = FALSE, warning = FALSE}
require(dplyr, quietly = T) #simple manipulation of data.frames
require(Seurat, quietly = T) #scRNA-seq analysis
require(reshape, quietly = T) #melting and casting data.frames
require(ggplot2, quietly = T) #plotting
require(rstatix, quietly = T) #adding p values to plots
require(ggprism, quietly = T) #adding p values to plots
require(ggsci, quietly = T) #IGV color palette
require(patchwork, quietly = T) #adding p values to plots
require(magrittr, quietly = T) #adding p values to plots
require(gridExtra, quietly = T) #display multiple plots
require(ggrepel, quietly = T) #labels for plots
require(tibble, quietly = T) #rownames_to_column function
require(useful, quietly = T) #corner function to peek at data.frames
require(viridis, quietly = T) #viridis color palettes
require(readxl, quietly = T) #reading Excel files
require(msigdbr, quietly = T) #accessing gene signatures from mSigDB
source("scripts/utils.R") #custom color palettes and other functions
set.seed(1) #set a seed for reproducibility
```

```{r read in Seurats}
seurat_integrated <- saveRDS("objects/seurat_integrated_clustered.rds")
pan4 <- readRDS("objects/seurat_cd4_panonly_clustered.rds")
pan8 <- readRDS("objects/seurat_cd8_panonly_clustered.rds")
```

First, UMAPs showing the distribution of PM and non-PM T cells in the blood will be plotted, followed by some pie charts illustrating the distribution of clusters for each group.
```{r UMAPs of shared/non-shared blood}
p_blood_shared <- seurat_integrated@meta.data %>% 
                    filter(tissue == "blood" & !is.na(pancreas_matching)) %>%
                    ggplot(aes(x = allcells_UMAP_1, 
                               y = allcells_UMAP_2, 
                               color = seurat_clusters)) + 
                    geom_point(data = seurat_integrated@meta.data %>% 
                                 dplyr::select(-pancreas_matching), 
                               shape = 16, 
                               color = "grey90") +
                    geom_point(shape = 16, 
                               alpha = 0.3) +
                    theme_minimal() +
                    theme(axis.title = element_blank(),
                          axis.text = element_blank(),
                          panel.grid = element_blank(),
                          text = element_text(size = 15),
                          legend.position = "none") +
                    labs(color=NULL,
                         title = "T cells in blood: pancreas-matching (PM) or non-matching") +
                    scale_color_manual(values = custom_colors("clusters")) +
                    guides(color = guide_legend(override.aes = list(alpha = 1, 
                                                                  size = 5, 
                                                                  shape = 19))) +
                  facet_grid(rows = vars(pancreas_matching), 
                             cols = vars(treatment))
p_blood_shared

#PIES OF CLUSTERS IN EACH GROUP
p_cd8_blood_pie <- seurat_integrated@meta.data %>% 
                    filter(tissue == "blood" & !is.na(pancreas_matching) & celltype == "CD8") %>%
                    group_by(treatment,pancreas_matching,
                             seurat_clusters) %>% 
                    tally() %>% 
                    group_by(treatment, pancreas_matching) %>% 
                    mutate(prcnt = n*100/sum(n),
                           csum = rev(cumsum(rev(prcnt))),  #get the position for the labels
                           pos = prcnt/2 + lead(csum, 1),
                           pos = if_else(is.na(pos), prcnt/2, pos)) %>%
                    ggplot(aes(x = "", 
                               y = prcnt, 
                               fill = seurat_clusters)) + # pct used here so slices add to 100
                    geom_bar(stat="identity", width=1) +
                    coord_polar("y", 
                                start = 0) +
                    #geom_text(aes(label = paste0(round(prcnt, 0), "%")), position = position_stack(vjust = 0.5)) +
                    facet_grid(rows = vars(pancreas_matching), 
                               cols = vars(treatment)) +
                    theme_void() +
                    theme(legend.position = "none",
                          text = element_text(size = 15)) +
                    labs(title = "CD8+ T cells in blood") +
                    # geom_label_repel(aes(y = pos, 
                    #                      label = paste0(round(prcnt, 0), "%")),
                    #                      size = 4.5, 
                    #                      nudge_x = 0.1, 
                    #                      show.legend = FALSE) +
                    scale_fill_manual(values = custom_colors("clusters"))
p_cd4_blood_pie <- seurat_integrated@meta.data %>% 
                    filter(tissue == "blood" & !is.na(pancreas_matching) & celltype %in% c("Tcon","Treg")) %>%
                    group_by(treatment,pancreas_matching,
                             seurat_clusters) %>% 
                    tally() %>% 
                    group_by(treatment, pancreas_matching) %>% 
                    mutate(prcnt = n*100/sum(n),
                           csum = rev(cumsum(rev(prcnt))),  #get the position for the labels
                           pos = prcnt/2 + lead(csum, 1),
                           pos = if_else(is.na(pos), prcnt/2, pos)) %>%
                    ggplot(aes(x = "", 
                               y = prcnt, 
                               fill = seurat_clusters)) + # pct used here so slices add to 100
                    geom_bar(stat="identity", width=1) +
                    coord_polar("y", 
                                start = 0) +
                    #geom_text(aes(label = paste0(round(prcnt, 0), "%")), position = position_stack(vjust = 0.5)) +
                    facet_grid(rows = vars(pancreas_matching), 
                               cols = vars(treatment)) +
                    theme_void() +
                    theme(legend.position = "none",
                          text = element_text(size = 15)) +
                    labs(title = "CD4+ T cells in blood") +
                    # geom_label_repel(aes(y = pos, 
                    #                      label = paste0(round(prcnt, 0), "%")),
                    #                      size = 4.5, 
                    #                      nudge_x = 0.1, 
                    #                      show.legend = FALSE) +
                    scale_fill_manual(values = custom_colors("clusters"))
grid.arrange(p_cd8_blood_pie, p_cd4_blood_pie, nrow = 2)
```

```{r diversity index for clusters in PM and non-PM}
## PANCREAS
#start with PM cells (CD8)
tcrtypecounts <- seurat_integrated@meta.data %>% 
    filter(tissue == "pancreas" & pancreas_matching == "PM" & cd4_or_cd8 == "CD8") %>%
    group_by(treatment,mouse_id,
             seurat_clusters) %>% 
    tally() %>% 
  reshape::cast(treatment+mouse_id~seurat_clusters, value= "n", fill = 0)
totalN <- tcrtypecounts %>% dplyr::select(-c(treatment, mouse_id)) %>% rowSums() # Total N by row sums
df <- sapply(tcrtypecounts %>% dplyr::select(-c(treatment, mouse_id)), function(x) (x - 1)*x) # Calculate x(x-1) for each value in df
corner(df)
numerator <- df %>% rowSums() # Calculate numerator of equation (x in above)
diversities <- data.frame(tcrtypecounts %>% 
                              dplyr::select(c(treatment, mouse_id)), 
                          celltype = "CD8", 
                          tissue = "pancreas",
                          pancreas_matching = "PM",
                          simpson = totalN*(totalN - 1) / numerator)

#non-PM cells (CD8)
tcrtypecounts <- seurat_integrated@meta.data %>% 
    filter(tissue == "pancreas" & pancreas_matching == "non-PM" & cd4_or_cd8 == "CD8") %>%
    group_by(treatment,mouse_id,
             seurat_clusters) %>% 
    tally() %>% 
  reshape::cast(treatment+mouse_id~seurat_clusters, value= "n", fill = 0)
totalN <- tcrtypecounts %>% dplyr::select(-c(treatment, mouse_id)) %>% rowSums() # Total N by row sums
df <- sapply(tcrtypecounts %>% dplyr::select(-c(treatment, mouse_id)), function(x) (x - 1)*x) # Calculate x(x-1) for each value in df
corner(df)
numerator <- df %>% rowSums() # Calculate numerator of equation (x in above)
diversities <- rbind(diversities, 
                                 data.frame(tcrtypecounts %>% 
                              dplyr::select(c(treatment, mouse_id)), 
                          celltype = "CD8", 
                          tissue = "pancreas",
                          pancreas_matching = "non-PM",
                          simpson = totalN*(totalN - 1) / numerator))

#non-PM cells (CD4)
tcrtypecounts <- seurat_integrated@meta.data %>% 
    filter(tissue == "pancreas" & pancreas_matching == "non-PM" & cd4_or_cd8 == "CD4") %>%
    group_by(treatment,mouse_id,
             seurat_clusters) %>% 
    tally() %>% 
  reshape::cast(treatment+mouse_id~seurat_clusters, value= "n", fill = 0)
totalN <- tcrtypecounts %>% dplyr::select(-c(treatment, mouse_id)) %>% rowSums() # Total N by row sums
df <- sapply(tcrtypecounts %>% dplyr::select(-c(treatment, mouse_id)), function(x) (x - 1)*x) # Calculate x(x-1) for each value in df
corner(df)
numerator <- df %>% rowSums() # Calculate numerator of equation (x in above)
diversities <- rbind(diversities, 
                                 data.frame(tcrtypecounts %>% 
                              dplyr::select(c(treatment, mouse_id)), 
                          celltype = "CD4", 
                          tissue = "pancreas",
                          pancreas_matching = "non-PM",
                          simpson = totalN*(totalN - 1) / numerator))

#PM cells (CD4)
tcrtypecounts <- seurat_integrated@meta.data %>% 
    filter(tissue == "pancreas" & pancreas_matching == "PM" & cd4_or_cd8 == "CD4") %>%
    group_by(treatment,mouse_id,
             seurat_clusters) %>% 
    tally() %>% 
  reshape::cast(treatment+mouse_id~seurat_clusters, value= "n", fill = 0)
totalN <- tcrtypecounts %>% dplyr::select(-c(treatment, mouse_id)) %>% rowSums() # Total N by row sums
df <- sapply(tcrtypecounts %>% dplyr::select(-c(treatment, mouse_id)), function(x) (x - 1)*x) # Calculate x(x-1) for each value in df
corner(df)
numerator <- df %>% rowSums() # Calculate numerator of equation (x in above)
diversities <- rbind(diversities, 
                                 data.frame(tcrtypecounts %>% 
                              dplyr::select(c(treatment, mouse_id)), 
                          celltype = "CD4", 
                          tissue = "pancreas",
                          pancreas_matching = "PM",
                          simpson = totalN*(totalN - 1) / numerator))

## PERIPHERY
#start with PM cells (CD8)
tcrtypecounts <- seurat_integrated@meta.data %>% 
    filter(tissue != "pancreas" & pancreas_matching == "PM" & cd4_or_cd8 == "CD8") %>%
    group_by(treatment,mouse_id,
             seurat_clusters) %>% 
    tally() %>% 
  reshape::cast(treatment+mouse_id~seurat_clusters, value= "n", fill = 0)
totalN <- tcrtypecounts %>% dplyr::select(-c(treatment, mouse_id)) %>% rowSums() # Total N by row sums
df <- sapply(tcrtypecounts %>% dplyr::select(-c(treatment, mouse_id)), function(x) (x - 1)*x) # Calculate x(x-1) for each value in df
corner(df)
numerator <- df %>% rowSums() # Calculate numerator of equation (x in above)
diversities <- rbind(diversities,
                     data.frame(tcrtypecounts %>% 
                              dplyr::select(c(treatment, mouse_id)), 
                          celltype = "CD8", 
                          tissue = "periphery",
                          pancreas_matching = "PM",
                          simpson = totalN*(totalN - 1) / numerator))

#non-PM cells (CD8)
tcrtypecounts <- seurat_integrated@meta.data %>% 
    filter(tissue != "pancreas" & pancreas_matching == "non-PM" & cd4_or_cd8 == "CD8") %>%
    group_by(treatment,mouse_id,
             seurat_clusters) %>% 
    tally() %>% 
  reshape::cast(treatment+mouse_id~seurat_clusters, value= "n", fill = 0)
totalN <- tcrtypecounts %>% dplyr::select(-c(treatment, mouse_id)) %>% rowSums() # Total N by row sums
df <- sapply(tcrtypecounts %>% dplyr::select(-c(treatment, mouse_id)), function(x) (x - 1)*x) # Calculate x(x-1) for each value in df
corner(df)
numerator <- df %>% rowSums() # Calculate numerator of equation (x in above)
diversities <- rbind(diversities, 
                                 data.frame(tcrtypecounts %>% 
                              dplyr::select(c(treatment, mouse_id)), 
                          celltype = "CD8", 
                          tissue = "periphery",
                          pancreas_matching = "non-PM",
                          simpson = totalN*(totalN - 1) / numerator))

#non-PM cells (CD4)
tcrtypecounts <- seurat_integrated@meta.data %>% 
    filter(tissue != "pancreas" & pancreas_matching == "non-PM" & cd4_or_cd8 == "CD4") %>%
    group_by(treatment,mouse_id,
             seurat_clusters) %>% 
    tally() %>% 
  reshape::cast(treatment+mouse_id~seurat_clusters, value= "n", fill = 0)
totalN <- tcrtypecounts %>% dplyr::select(-c(treatment, mouse_id)) %>% rowSums() # Total N by row sums
df <- sapply(tcrtypecounts %>% dplyr::select(-c(treatment, mouse_id)), function(x) (x - 1)*x) # Calculate x(x-1) for each value in df
corner(df)
numerator <- df %>% rowSums() # Calculate numerator of equation (x in above)
diversities <- rbind(diversities, 
                                 data.frame(tcrtypecounts %>% 
                              dplyr::select(c(treatment, mouse_id)), 
                          celltype = "CD4", 
                          tissue = "periphery",
                          pancreas_matching = "non-PM",
                          simpson = totalN*(totalN - 1) / numerator))

#PM cells (CD4)
tcrtypecounts <- seurat_integrated@meta.data %>% 
    filter(tissue != "pancreas" & pancreas_matching == "PM" & cd4_or_cd8 == "CD4") %>%
    group_by(treatment,mouse_id,
             seurat_clusters) %>% 
    tally() %>% 
  reshape::cast(treatment+mouse_id~seurat_clusters, value= "n", fill = 0)
totalN <- tcrtypecounts %>% dplyr::select(-c(treatment, mouse_id)) %>% rowSums() # Total N by row sums
df <- sapply(tcrtypecounts %>% dplyr::select(-c(treatment, mouse_id)), function(x) (x - 1)*x) # Calculate x(x-1) for each value in df
corner(df)
numerator <- df %>% rowSums() # Calculate numerator of equation (x in above)
diversities <- rbind(diversities, 
                                 data.frame(tcrtypecounts %>% 
                              dplyr::select(c(treatment, mouse_id)), 
                          celltype = "CD4", 
                          tissue = "periphery",
                          pancreas_matching = "PM",
                          simpson = totalN*(totalN - 1) / numerator))


diversities$treatment <- factor(diversities$treatment, levels = c("IgG", "Spt", "PD1")) #factor the pathologies

p_periphery <- diversities %>%
    filter(pancreas_matching == "non-PM") %>%
    ggplot(aes(x = treatment,
               y = simpson,
               fill = treatment,
               color = treatment)) +
    geom_violin(alpha = 0.5,
                adjust = 1,
                scale = "width",
                draw_quantiles = 0.5) +
    geom_point(size = 4) +
    labs(y = "Diversity index") +
    theme_minimal() +
    theme(panel.grid = element_blank(),
          text = element_text(size = 15),
          panel.border = element_rect(colour = "black",
                                      fill=NA,
                                      size=1),
          #strip.text = element_blank(),
          axis.ticks.y = element_line(),
          panel.spacing = unit(0, "cm"),
          legend.position = "none") +
    labs(x = NULL) +
    scale_fill_manual(values = custom_colors("treatments")) +
    scale_color_manual(values = custom_colors("treatments")) +
    scale_y_continuous(limits = c(1,9), breaks = c(2,4,6,8)) +
    facet_wrap(tissue ~ celltype)

p_pancreas <- diversities %>%
    filter(pancreas_matching == "PM") %>%
    ggplot(aes(x = treatment,
               y = simpson,
               fill = treatment,
               color = treatment)) +
    geom_violin(alpha = 0.5,
                adjust = 1,
                scale = "width",
                draw_quantiles = 0.5) +
    geom_point(size = 4) +
    labs(y = "Diversity index") +
    theme_minimal() +
    theme(panel.grid = element_blank(),
          text = element_text(size = 15),
          panel.border = element_rect(colour = "black",
                                      fill=NA,
                                      size=1),
          #strip.text = element_blank(),
          axis.ticks.y = element_line(),
          panel.spacing = unit(0, "cm"),
          legend.position = "none") +
    labs(x = NULL) +
    scale_fill_manual(values = custom_colors("treatments")) +
    scale_color_manual(values = custom_colors("treatments")) +
    scale_y_continuous(limits = c(1,9), breaks = c(2,4,6,8)) +
    facet_wrap(tissue ~ celltype)

grid.arrange(p_pancreas, p_periphery, nrow = 1)
```

Show the top 50 clonotypes as a percentage of TCRs detected in the PANCREAS as a cumulative sum and also the clone sizes for the different treatment groups.
```{r top 50 clonotypes and clone size in pancreas}
#TOP 50 CLONOTYPES: CD8 IN PANCREAS
cumulative_sums <- seurat_integrated@meta.data %>% 
                    filter(clonotype_tissue_groups %in% c("blood&pancreas","pancreas") & celltype == "CD8") %>%
                      group_by(treatment, clonotype_id, pancreas_matching) %>% 
                      tally() %>% 
                      group_by(treatment) %>% 
                      mutate(prcnt = n*100/sum(n)) %>% 
                      group_by(treatment, pancreas_matching) %>% 
                      arrange(treatment, pancreas_matching, desc(prcnt)) %>% 
                      mutate(cs = cumsum(prcnt))

cumulative_sums$clonotype_id <- factor(cumulative_sums$clonotype_id, 
                                       levels = cumulative_sums$clonotype_id) #orders the clonotypes on x axis

p_cd8_panPM_top50_clonos <- cumulative_sums %>% 
                              group_by(treatment, pancreas_matching) %>% 
                              slice_head(n = 50) %>% 
                              ggplot(aes(x = clonotype_id, 
                                         y = cs, 
                                         color = treatment)) + 
                              geom_point() + 
                              geom_segment(aes(xend = clonotype_id, 
                                               yend = cs, 
                                               x = clonotype_id, 
                                               y = 0, 
                                               color = treatment)) + 
                              scale_color_manual(values = custom_colors("treatments")) +
                              scale_y_continuous(expand = c(0, 0)) +
                              theme_minimal() +
                              theme(panel.grid = element_blank(),
                                    panel.spacing.x = unit(0, "lines"),
                                    text = element_text(size = 15),
                                    axis.text.x = element_blank(),
                                    panel.border = element_rect(colour = "black", 
                                                                fill = NA, 
                                                                size = 1),
                                    legend.position = "none") +
                              labs(x = "Top 50 clonotypes", 
                                   y = "Cumulative % CD8 TCRs",
                                   title = "CD8+ T cells in pancreas") +
                              facet_wrap(treatment ~ pancreas_matching, scales = "free_x", ncol = 2)

#TOP 50 CLONOTYPES: TCON IN PANCREAS
cumulative_sums <- seurat_integrated@meta.data %>% 
                    filter(clonotype_tissue_groups %in% c("blood&pancreas","pancreas") & celltype == "Tcon") %>%
                      group_by(treatment, clonotype_id, pancreas_matching) %>% 
                      tally() %>% 
                      group_by(treatment) %>% 
                      mutate(prcnt = n*100/sum(n)) %>% 
                      group_by(treatment, pancreas_matching) %>% 
                      arrange(treatment, pancreas_matching, desc(prcnt)) %>% 
                      mutate(cs = cumsum(prcnt))

cumulative_sums$clonotype_id <- factor(cumulative_sums$clonotype_id, 
                                       levels = cumulative_sums$clonotype_id) #orders the clonotypes on x axis

p_tcon_panPM_top50_clonos <- cumulative_sums %>% 
                              group_by(treatment, pancreas_matching) %>% 
                              slice_head(n = 50) %>% 
                              ggplot(aes(x = clonotype_id, 
                                         y = cs, 
                                         color = treatment)) + 
                              geom_point() + 
                              geom_segment(aes(xend = clonotype_id, 
                                               yend = cs, 
                                               x = clonotype_id, 
                                               y = 0, 
                                               color = treatment)) + 
                              scale_color_manual(values = custom_colors("treatments")) +
                              scale_y_continuous(expand = c(0, 0)) +
                              theme_minimal() +
                              theme(panel.grid = element_blank(),
                                    panel.spacing.x = unit(0, "lines"),
                                    text = element_text(size = 15),
                                    axis.text.x = element_blank(),
                                    panel.border = element_rect(colour = "black", 
                                                                fill = NA, 
                                                                size = 1),
                                    legend.position = "none") +
                              labs(x = "Top 50 clonotypes", 
                                   y = "Cumulative % Tcon TCRs",
                                   title = "Tcon in pancreas") +
                              facet_wrap(treatment ~ pancreas_matching, scales = "free_x", ncol = 2)

#TOP 50 CLONOTYPES: TREG IN PANCREAS
cumulative_sums <- seurat_integrated@meta.data %>% 
                    filter(clonotype_tissue_groups %in% c("blood&pancreas","pancreas") & celltype == "Treg") %>%
                      group_by(treatment, clonotype_id, pancreas_matching) %>% 
                      tally() %>% 
                      group_by(treatment) %>% 
                      mutate(prcnt = n*100/sum(n)) %>% 
                      group_by(treatment, pancreas_matching) %>% 
                      arrange(treatment, pancreas_matching, desc(prcnt)) %>% 
                      mutate(cs = cumsum(prcnt))

cumulative_sums$clonotype_id <- factor(cumulative_sums$clonotype_id, 
                                       levels = cumulative_sums$clonotype_id) #orders the clonotypes on x axis

p_treg_panPM_top50_clonos <- cumulative_sums %>% 
                              group_by(treatment, pancreas_matching) %>% 
                              slice_head(n = 50) %>% 
                              ggplot(aes(x = clonotype_id, 
                                         y = cs, 
                                         color = treatment)) + 
                              geom_point() + 
                              geom_segment(aes(xend = clonotype_id, 
                                               yend = cs, 
                                               x = clonotype_id, 
                                               y = 0, 
                                               color = treatment)) + 
                              scale_color_manual(values = custom_colors("treatments")) +
                              scale_y_continuous(expand = c(0, 0)) +
                              theme_minimal() +
                              theme(panel.grid = element_blank(),
                                    panel.spacing.x = unit(0, "lines"),
                                    text = element_text(size = 15),
                                    axis.text.x = element_blank(),
                                    panel.border = element_rect(colour = "black", 
                                                                fill = NA, 
                                                                size = 1),
                                    legend.position = "none") +
                              labs(x = "Top 50 clonotypes", 
                                   y = "Cumulative % Treg TCRs",
                                   title = "Treg in pancreas") +
                              facet_wrap(treatment ~ pancreas_matching, scales = "free_x", ncol = 2)

grid.arrange(p_cd8_panPM_top50_clonos, p_tcon_panPM_top50_clonos, p_treg_panPM_top50_clonos, nrow = 1)

#BOXPLOT OF CLONE SIZE
stat.test <- rbind(data.frame(seurat_integrated@meta.data %>%
                          filter(tissue == "pancreas" & !is.na(pancreas_matching) & celltype == "CD8") %>%
                          group_by(treatment, pancreas_matching, clonotype_id) %>%
                          tally() %>%
                          group_by(pancreas_matching) %>%
                          wilcox_test(n ~ treatment) %>% 
                          add_xy_position() %>%
                          mutate(y.position = log10(y.position + 1),
                                 celltype = "CD8")),
                   data.frame(seurat_integrated@meta.data %>%
                          filter(tissue == "pancreas" & !is.na(pancreas_matching) & celltype == "Tcon") %>%
                          group_by(treatment, pancreas_matching, clonotype_id) %>%
                          tally() %>%
                          group_by(pancreas_matching) %>%
                          wilcox_test(n ~ treatment) %>% 
                          add_xy_position() %>%
                          mutate(y.position = log10(y.position + 1),
                                 celltype = "Tcon")))

p_boxplot_pm_pan <- seurat_integrated@meta.data %>%
                      filter(tissue == "pancreas" & !is.na(pancreas_matching) & celltype != "Treg") %>%
                      group_by(celltype, treatment, pancreas_matching, clonotype_id) %>%
                      tally() %>%
                      ggplot() + 
                      geom_boxplot(aes(x = pancreas_matching, 
                                 y = log10(n), 
                                 fill = treatment,
                                 color = treatment),
                                 outlier.shape = NA, 
                                 alpha = 0.5,
                                 position = position_dodge(0.9)) +
                      geom_point(aes(x = pancreas_matching, 
                                 y = log10(n), 
                                 fill = treatment,
                                 color = treatment),
                                 position = position_jitterdodge(dodge.width = 0.9,
                                                                 jitter.width = 0.4),
                                 alpha = 0.1) +
                      scale_fill_manual(values = custom_colors("treatments")) +
                      scale_color_manual(values = custom_colors("treatments")) +
                      theme_minimal() +
                      theme(panel.grid = element_blank(),
                            axis.title.x = element_blank(),
                            text = element_text(),
                            panel.spacing = unit(0, "cm"),
                            legend.position = "top",
                            panel.border = element_rect(colour = "black", fill = NA, size = 1)) +
                      labs(y = expression(log[10](clone~size)),
                           color = NULL,
                           fill = NULL,
                           title = "T cells in pancreas (unique clonotypes)") +
                      facet_wrap(~celltype, nrow = 1)

#DIVERSITY ANALYSIS
tcrtypecounts <- rbind(data.frame(pan8@meta.data %>% dplyr::select(treatment, mouse_id, clonotype_id, pancreas_matching, pan_clonotype_count)))
corner(tcrtypecounts)

tcrtypecounts <- tcrtypecounts %>% filter(!is.na(clonotype_id)) %>% cast(treatment + mouse_id + pancreas_matching  ~ clonotype_id)
corner(tcrtypecounts)

totalN <- tcrtypecounts %>% select(-c(treatment, mouse_id, pancreas_matching)) %>% rowSums() # Total N by row sums
df <- sapply(tcrtypecounts %>% select(-c(treatment, mouse_id, pancreas_matching)), function(x) (x - 1)*x) # Calculate x(x-1) for each value in df
corner(df)
numerator <- df %>% rowSums() # Calculate numerator of equation (x in above)

# Calculate simpson index
diversities <- data.frame(tcrtypecounts %>%
                            dplyr::select(c(treatment, mouse_id, pancreas_matching)),
                          celltype = "CD8",
                          simpson = totalN*(totalN - 1) / numerator)

diversities$treatment <- factor(diversities$treatment, levels = c("IgG", "Spt", "PD1")) #factor the pathologies

stat.test <- data.frame(diversities %>%
                            mutate(simpson = ifelse(simpson == Inf, NA, simpson)) %>%
                          group_by(pancreas_matching) %>%
                            wilcox_test(simpson ~ treatment) %>%
                            add_xy_position())

p_diversities_pm_pan <- diversities %>%
    ggplot(aes(x = treatment,
               y = simpson,
               fill = treatment,
               color = treatment)) +
    geom_violin(alpha = 0.5,
                adjust = 1,
                scale = "width",
                trim = F) +
    geom_point(size = 5) +
    labs(y = "Simpson index") +
    theme_minimal() +
    theme(panel.grid = element_blank(),
          text = element_text(size = 15),
          panel.border = element_rect(colour = "black",
                                      fill=NA,
                                      size=1),
          legend.position = "bottom") +
    labs(title = "CD8 T cells in pancreas") +
    scale_fill_manual(values = custom_colors("treatments")) +
    scale_color_manual(values = custom_colors("treatments")) +
    add_pvalue(stat.test,
               inherit.aes = F,
               label = "p.adj.signif",
                y.position = "y.position") +
    facet_wrap(~pancreas_matching, scales = "free_y")

```

```{r upset plot and heatmap of shared clusters in blood PM}
cluster_labels <- levels(factor(pan8$seurat_clusters))
heatmap_df <- data.frame(matrix(ncol = length(cluster_labels), nrow = length(cluster_labels)), row.names = cluster_labels)
colnames(heatmap_df) <- cluster_labels

clonotype_sharing <- seurat_integrated@meta.data %>%
                        filter(!is.na(clonotype_id)) %>%
                        dplyr::select(treatment,clonotype_id,seurat_clusters) %>%
                        unique() %>% mutate(index = TRUE) %>%
                        pivot_wider(c(treatment, clonotype_id), names_from = seurat_clusters, values_from = index, values_fill = FALSE)

for (i in 1:(length(cluster_labels)-1)) {
  for (j in (i+1):length(cluster_labels)) {
    filter_string <- paste0(cluster_labels[i], " & ", cluster_labels[j])
    heatmap_df[i,j] <- clonotype_sharing %>%
                        filter(!! rlang::parse_expr(filter_string)) %>% 
                        nrow()
  }
}


clonos_of_interest <- seurat_integrated@meta.data %>% 
                        filter(treatment == "PD1" & seurat_clusters == "CD8_texh" & pancreas_matching == "PM") %>% 
                        pull(clonotype_id)
df <- seurat_integrated@meta.data %>% 
          filter(clonotype_id %in% clonos_of_interest) %>% 
          group_by(treatment,mouse_id,clonotype_id,clonotype_count,seurat_clusters) %>% 
          unique() %>% 
          dplyr::select(treatment,mouse_id,clonotype_id,clonotype_count,seurat_clusters) %>% 
          summarize() %>%
          mutate(index = T) %>% 
          pivot_wider(c(treatment,mouse_id,clonotype_id,clonotype_count), names_from = seurat_clusters, values_from = index) %>% 
          mutate_all(replace_na, F)
cluster_labels <- rev(levels(factor(pan8$seurat_clusters))) #determines the order of the clusters in the table
df$num_unique_clusters <- rowSums(df[,5:15])
df$grouping <- apply(df, 1, function(x) paste(colnames(df)[5:15][x[5:15] == T], collapse = "&")) #add a column that concatenates all the seurat clusters that a clonotype includes
upset(df, cluster_labels, set_sizes = F, sort_sets = F, width_ratio = 0.1, min_size = 2,  keep_empty_groups=TRUE) +
    ggtitle("Terminally exhausted/effector-like shared") + 
  theme_minimal() +
  theme(text = element_text(size = 15))

clonos_of_interest <- pan8@meta.data %>% 
                        filter(pancreas_matching == "PM" & treatment == "Spt" & seurat_clusters == "CD8_pexh") %>% 
                        pull(clonotype_id)
df <- seurat_integrated@meta.data %>% 
          filter(clonotype_id %in% clonos_of_interest) %>% 
          group_by(treatment,mouse_id,clonotype_id,clonotype_count,seurat_clusters) %>% 
          unique() %>% 
          dplyr::select(treatment,mouse_id,clonotype_id,clonotype_count,seurat_clusters) %>% 
          summarize() %>%
          mutate(index = T) %>% 
          pivot_wider(c(treatment,mouse_id,clonotype_id,clonotype_count), names_from = seurat_clusters, values_from = index) %>% 
          mutate_all(replace_na, F)
cluster_labels <- rev(levels(factor(pan8$seurat_clusters))) #determines the order of the clusters in the table
df$num_unique_clusters <- rowSums(df[,5:14])
df$grouping <- apply(df, 1, function(x) paste(colnames(df)[5:14][x[5:14] == T], collapse = "&")) #add a column that concatenates all the seurat clusters that a clonotype includes
upset(df, cluster_labels[-11], set_sizes = F, sort_sets = F, width_ratio = 0.1, min_size = 2,  keep_empty_groups=TRUE) +
    ggtitle("Terminally exhausted/effector-like shared") + 
  theme_minimal() +
  theme(text = element_text(size = 15))
```

```{r percent blood PM matching}
rbind(seurat_integrated@meta.data %>% 
          filter(!is.na(pancreas_matching) & cd4_or_cd8 == "CD4" & tissue == "blood") %>% 
          group_by(treatment, mouse_id, pancreas_matching) %>% 
          tally() %>% 
          mutate(celltype="CD4") %>% 
          cast(treatment+mouse_id+celltype ~ pancreas_matching, value = "n"),
      seurat_integrated@meta.data %>% 
          filter(!is.na(pancreas_matching) & cd4_or_cd8 == "CD8" & tissue == "blood") %>% 
          group_by(treatment, mouse_id, pancreas_matching) %>% 
          tally() %>% 
          mutate(celltype="CD8") %>% 
          cast(treatment+mouse_id+celltype ~ pancreas_matching, value = "n")) %>% 
    group_by(treatment, mouse_id, celltype) %>% 
    summarize(ratio = PM*100/(`non-PM`+PM)) %>% 
    ggplot(aes(x = treatment, 
               color = treatment, 
               fill = treatment, 
               y = ratio)) + 
    geom_boxplot(position = position_dodge(0.8),alpha=0.4) + 
    geom_point(position = position_dodge(0.8),size=4) + 
    theme_minimal() +
    theme(panel.grid = element_blank(),
          panel.border = element_rect(color= "black", fill =NA, size=0.5),
          axis.title.x = element_blank(),
          text = element_text(size=15)) +
    scale_color_manual(values = custom_colors("treatments")) +
    scale_fill_manual(values = custom_colors("treatments")) +
    facet_wrap(~ celltype, 
               scales = "free_y") +
    labs(y = "PM (%)",
         title = "T cells in blood")

rbind(seurat_integrated@meta.data %>% 
          filter(!is.na(pancreas_matching) & cd4_or_cd8 == "CD4" & tissue == "blood") %>% 
          group_by(treatment, mouse_id, pancreas_matching) %>% 
          tally() %>% 
          mutate(celltype="CD4") %>% 
          cast(treatment+mouse_id+celltype ~ pancreas_matching, value = "n"),
      seurat_integrated@meta.data %>% 
          filter(!is.na(pancreas_matching) & cd4_or_cd8 == "CD8" & tissue == "blood" & seurat_clusters != "CD8_cm") %>% 
          group_by(treatment, mouse_id, pancreas_matching) %>% 
          tally() %>% 
          mutate(celltype="CD8") %>% 
          cast(treatment+mouse_id+celltype ~ pancreas_matching, value = "n")) %>% 
    group_by(treatment, mouse_id, celltype) %>% 
    summarize(ratio = PM*100/(`non-PM`+PM)) %>% 
    ggplot(aes(x = treatment, 
               color = treatment, 
               fill = treatment, 
               y = ratio)) + 
    geom_boxplot(position = position_dodge(0.8),alpha=0.4) + 
    geom_point(position = position_dodge(0.8),size=4) + 
    theme_minimal() +
    theme(panel.grid = element_blank(),
          panel.border = element_rect(color= "black", fill =NA, size=0.5),
          axis.title.x = element_blank(),
          text = element_text(size=15)) +
    scale_color_manual(values = custom_colors("treatments")) +
    scale_fill_manual(values = custom_colors("treatments")) +
    facet_wrap(~ celltype, 
               scales = "free_y") +
    labs(y = "PM (%)",
         title = "T cells in blood")
```

```{r PM in NRP-v7+ tetramer-specific T cells}

seurat_integrated@meta.data %>% 
    filter(!is.na(pancreas_matching)) %>% 
    mutate(andreas_tetramer_specificity = ifelse(is.na(andreas_tetramer_specificity), "unknown", "NRP-v7")) %>%
    group_by(andreas_tetramer_specificity, pancreas_matching) %>% 
    tally() %>% 
    ggplot(aes(x=andreas_tetramer_specificity, y=n, fill = pancreas_matching)) + 
    geom_bar(stat="identity", position = "fill") +
    theme_minimal() +
    theme(panel.grid = element_blank(),
          axis.title.x = element_blank()) +
    scale_fill_manual(values = custom_colors("pm")) + 
  coord_flip()
```

Show the top 50 clonotypes as a percentage of TCRs detected in the BLOOD as a cumulative sum and also the clone sizes for the different treatment groups.
```{r top 50 clonotypes and clone size in blood}
#TOP 50 CLONOTYPES: CD8 IN BLOOD
cumulative_sums <- seurat_integrated@meta.data %>% 
                    filter(clonotype_tissue_groups %in% c("blood&pancreas","blood") & celltype == "CD8") %>%
                      group_by(treatment, clonotype_id, pancreas_matching) %>% 
                      tally() %>% 
                      group_by(treatment) %>% 
                      mutate(prcnt = n*100/sum(n)) %>% 
                      group_by(treatment, pancreas_matching) %>% 
                      arrange(treatment, pancreas_matching, desc(prcnt)) %>% 
                      mutate(cs = cumsum(prcnt))

cumulative_sums$clonotype_id <- factor(cumulative_sums$clonotype_id, 
                                       levels = cumulative_sums$clonotype_id) #orders the clonotypes on x axis

p_cd8_bldPM_top50_clonos <- cumulative_sums %>% 
                              group_by(treatment, pancreas_matching) %>% 
                              slice_head(n = 50) %>% 
                              ggplot(aes(x = clonotype_id, 
                                         y = cs, 
                                         color = treatment)) + 
                              geom_point() + 
                              geom_segment(aes(xend = clonotype_id, 
                                               yend = cs, 
                                               x = clonotype_id, 
                                               y = 0, 
                                               color = treatment)) + 
                              scale_color_manual(values = custom_colors("treatments")) +
                              scale_y_continuous(expand = c(0, 0)) +
                              theme_minimal() +
                              theme(panel.grid = element_blank(),
                                    panel.spacing.x = unit(0, "lines"),
                                    text = element_text(size = 15),
                                    axis.text.x = element_blank(),
                                    panel.border = element_rect(colour = "black", 
                                                                fill = NA, 
                                                                size = 1),
                                    legend.position = "none") +
                              labs(x = "Top 50 clonotypes", 
                                   y = "Cumulative % CD8 TCRs",
                                   title = "CD8+ T cells in blood") +
                              facet_wrap(treatment ~ pancreas_matching, scales = "free_x", ncol = 2)

#TOP 50 CLONOTYPES: TCON IN BLOOD
cumulative_sums <- seurat_integrated@meta.data %>% 
                    filter(clonotype_tissue_groups %in% c("blood&pancreas","blood") & celltype == "Tcon") %>%
                      group_by(treatment, clonotype_id, pancreas_matching) %>% 
                      tally() %>% 
                      group_by(treatment) %>% 
                      mutate(prcnt = n*100/sum(n)) %>% 
                      group_by(treatment, pancreas_matching) %>% 
                      arrange(treatment, pancreas_matching, desc(prcnt)) %>% 
                      mutate(cs = cumsum(prcnt))

cumulative_sums$clonotype_id <- factor(cumulative_sums$clonotype_id, 
                                       levels = cumulative_sums$clonotype_id) #orders the clonotypes on x axis

p_tcon_bldPM_top50_clonos <- cumulative_sums %>% 
                              group_by(treatment, pancreas_matching) %>% 
                              slice_head(n = 50) %>% 
                              ggplot(aes(x = clonotype_id, 
                                         y = cs, 
                                         color = treatment)) + 
                              geom_point() + 
                              geom_segment(aes(xend = clonotype_id, 
                                               yend = cs, 
                                               x = clonotype_id, 
                                               y = 0, 
                                               color = treatment)) + 
                              scale_color_manual(values = custom_colors("treatments")) +
                              scale_y_continuous(expand = c(0, 0)) +
                              theme_minimal() +
                              theme(panel.grid = element_blank(),
                                    panel.spacing.x = unit(0, "lines"),
                                    text = element_text(size = 15),
                                    axis.text.x = element_blank(),
                                    panel.border = element_rect(colour = "black", 
                                                                fill = NA, 
                                                                size = 1),
                                    legend.position = "none") +
                              labs(x = "Top 50 clonotypes", 
                                   y = "Cumulative % Tcon TCRs",
                                   title = "Tcon in blood") +
                              facet_wrap(treatment ~ pancreas_matching, scales = "free_x", ncol = 2)

#TOP 50 CLONOTYPES: TREG IN PANCREAS
cumulative_sums <- seurat_integrated@meta.data %>% 
                    filter(clonotype_tissue_groups %in% c("blood&pancreas","blood") & celltype == "Treg") %>%
                      group_by(treatment, clonotype_id, pancreas_matching) %>% 
                      tally() %>% 
                      group_by(treatment) %>% 
                      mutate(prcnt = n*100/sum(n)) %>% 
                      group_by(treatment, pancreas_matching) %>% 
                      arrange(treatment, pancreas_matching, desc(prcnt)) %>% 
                      mutate(cs = cumsum(prcnt))

cumulative_sums$clonotype_id <- factor(cumulative_sums$clonotype_id, 
                                       levels = cumulative_sums$clonotype_id) #orders the clonotypes on x axis

p_treg_bldPM_top50_clonos <- cumulative_sums %>% 
                              group_by(treatment, pancreas_matching) %>% 
                              slice_head(n = 50) %>% 
                              ggplot(aes(x = clonotype_id, 
                                         y = cs, 
                                         color = treatment)) + 
                              geom_point() + 
                              geom_segment(aes(xend = clonotype_id, 
                                               yend = cs, 
                                               x = clonotype_id, 
                                               y = 0, 
                                               color = treatment)) + 
                              scale_color_manual(values = custom_colors("treatments")) +
                              scale_y_continuous(expand = c(0, 0)) +
                              theme_minimal() +
                              theme(panel.grid = element_blank(),
                                    panel.spacing.x = unit(0, "lines"),
                                    text = element_text(size = 15),
                                    axis.text.x = element_blank(),
                                    panel.border = element_rect(colour = "black", 
                                                                fill = NA, 
                                                                size = 1),
                                    legend.position = "none") +
                              labs(x = "Top 50 clonotypes", 
                                   y = "Cumulative % Treg TCRs",
                                   title = "Treg in blood") +
                              facet_wrap(treatment ~ pancreas_matching, scales = "free_x", ncol = 2)

grid.arrange(p_cd8_bldPM_top50_clonos, p_tcon_bldPM_top50_clonos, p_treg_bldPM_top50_clonos, nrow = 1)

#BOXPLOT OF CLONE SIZE
stat.test <- rbind(data.frame(seurat_integrated@meta.data %>%
                          filter(tissue == "blood" & !is.na(pancreas_matching) & celltype == "CD8") %>%
                          group_by(treatment, pancreas_matching, clonotype_id) %>%
                          tally() %>%
                          group_by(pancreas_matching) %>%
                          wilcox_test(n ~ treatment) %>% 
                          add_xy_position() %>%
                          mutate(y.position = log10(y.position + 1),
                                 celltype = "CD8")),
                   data.frame(seurat_integrated@meta.data %>%
                          filter(tissue == "blood" & !is.na(pancreas_matching) & celltype == "Tcon") %>%
                          group_by(treatment, pancreas_matching, clonotype_id) %>%
                          tally() %>%
                          group_by(pancreas_matching) %>%
                          wilcox_test(n ~ treatment) %>% 
                          add_xy_position() %>%
                          mutate(y.position = log10(y.position + 1),
                                 celltype = "Tcon")))

p_boxplot_pm_bld <- seurat_integrated@meta.data %>%
                      filter(tissue == "blood" & !is.na(pancreas_matching) & celltype != "Treg") %>%
                      group_by(celltype, treatment, pancreas_matching, clonotype_id) %>%
                      tally() %>% 
                      ggplot() + 
                      geom_boxplot(aes(x = pancreas_matching, 
                                 y = log10(n), 
                                 fill = treatment,
                                 color = treatment),
                                 outlier.shape = NA, 
                                 alpha = 0.5,
                                 position = position_dodge(0.9)) +
                      geom_point(aes(x = pancreas_matching, 
                                 y = log10(n), 
                                 fill = treatment,
                                 color = treatment),
                                 position = position_jitterdodge(dodge.width = 0.9,
                                                                 jitter.width = 0.4),
                                 alpha = 0.1) +
                      scale_fill_manual(values = custom_colors("treatments")) +
                      scale_color_manual(values = custom_colors("treatments")) +
                      theme_minimal() +
                      theme(panel.grid = element_blank(),
                            axis.title.x = element_blank(),
                            text = element_text(),
                             panel.spacing = unit(0, "cm"),
                            legend.position = "top",
                            panel.border = element_rect(colour = "black", fill = NA, size = 1)) +
                      labs(y = expression(log[10](clone~size)),
                           color = NULL,
                           fill = NULL,
                           title = "T cells in blood (unique clonotypes)") +
                      facet_wrap(~celltype, nrow = 1)

#DIVERSITY ANALYSIS
tcrtypecounts <- rbind(data.frame(seurat_integrated@meta.data %>% filter(tissue == "blood") %>% dplyr::select(treatment, mouse_id, clonotype_id, pancreas_matching, pan_clonotype_count)))
corner(tcrtypecounts)

tcrtypecounts <- tcrtypecounts %>% filter(!is.na(clonotype_id)) %>% cast(treatment + mouse_id + pancreas_matching  ~ clonotype_id)
corner(tcrtypecounts)

totalN <- tcrtypecounts %>% select(-c(treatment, mouse_id, pancreas_matching)) %>% rowSums() # Total N by row sums
df <- sapply(tcrtypecounts %>% select(-c(treatment, mouse_id, pancreas_matching)), function(x) (x - 1)*x) # Calculate x(x-1) for each value in df
corner(df)
numerator <- df %>% rowSums() # Calculate numerator of equation (x in above)

# Calculate simpson index
diversities <- data.frame(tcrtypecounts %>%
                            dplyr::select(c(treatment, mouse_id, pancreas_matching)),
                          celltype = "CD8",
                          simpson = totalN*(totalN - 1) / numerator)

diversities$treatment <- factor(diversities$treatment, levels = c("IgG", "Spt", "PD1")) #factor the pathologies

stat.test <- data.frame(diversities %>%
                            mutate(simpson = ifelse(simpson == Inf, NA, simpson)) %>%
                          group_by(pancreas_matching) %>%
                            wilcox_test(simpson ~ treatment) %>%
                            add_xy_position())

p_diversities_pm_bld <- diversities %>%
    ggplot(aes(x = treatment,
               y = simpson,
               fill = treatment,
               color = treatment)) +
    geom_violin(alpha = 0.5,
                adjust = 1,
                scale = "width",
                trim = F) +
    geom_point(size = 5) +
    labs(y = "Simpson index") +
    theme_minimal() +
    theme(panel.grid = element_blank(),
          text = element_text(size = 15),
          panel.border = element_rect(colour = "black",
                                      fill=NA,
                                      size=1),
          legend.position = "bottom") +
    labs(title = "CD8 T cells in blood") +
    scale_fill_manual(values = custom_colors("treatments")) +
    scale_color_manual(values = custom_colors("treatments")) +
    facet_wrap(~pancreas_matching, scales = "free_y") +
      add_pvalue(stat.test,
               inherit.aes = F,
               label = "p.adj.signif",
                y.position = "y.position")

```

Plot the correlation of clonotype sizes in both pancreas and blood
```{r correlation of clonotype size in pancreas and blood}
seurat_integrated@meta.data %>% 
    filter(clonotype_tissue_groups %in% c("blood&pancreas","pLN&blood&pancreas") & tissue %in% c("blood","pancreas")) %>%
    group_by(treatment, cd4_or_cd8, clonotype_id, tissue) %>% 
    tally() %>% 
    cast(treatment+cd4_or_cd8+clonotype_id~tissue, value = "n") %>%
    group_by(treatment, cd4_or_cd8, blood, pancreas) %>%
    tally(name = "number_of_clonotypes") %>%
    ggplot(aes(x = log10(pancreas),
               y = log10(blood),
               size = number_of_clonotypes,
               color = treatment)) +
    geom_point() +
    theme_minimal() +
    facet_wrap(cd4_or_cd8~treatment) + 
    scale_color_manual(values = custom_colors("treatments")) +
    theme_minimal() +
    theme(panel.grid = element_blank(),
          text = element_text(size = 15),
          panel.spacing.x = unit(0, "lines"),
          legend.position = "right",
          panel.border = element_rect(colour = "black", fill = NA, size = 1)) +
    labs(y = expression(log[10](blood~clone~size)),
         x = expression(log[10](pancreas~clone~size)),
         color = "Treatment",
         title = "T cells in blood",
         size = "Number of\nunique clonotypes") +
  scale_size_continuous(range = c(2,10))
```

DE genes volcano plot within pancreas matching cells in blood:
```{r volcano plot of PM T cells in blood}
cutoff <- 0.25 #set a cutoff for DE gene expression
ident1 <- "PD1"
ident2 <- "IgG"

Idents(seurat_integrated) <- seurat_integrated$tissue
seurat_bld <- subset(seurat_integrated, idents = "blood") #subset on T cells in blood
Idents(seurat_bld) <- seurat_bld$pancreas_matching
seurat_bld <- subset(seurat_bld, idents = "PM") #subset on PM T cells only
Idents(seurat_bld) <- seurat_bld$celltype

volcanodata <- list()
allvolcano <- data.frame()
for (current_cell in c("CD8", "Tcon", "Treg")) {
  volcanodata[[current_cell]] <- FindMarkers(seurat_bld, #define the table of DE genes
                                     ident.1 = ident1,
                                     ident.2 = ident2, 
                                     verbose = F,
                                     group.by = "treatment",
                                     subset.ident = current_cell,
                                     only.pos = F, 
                                     logfc.threshold = cutoff, 
                                     assay = "RNA") %>% 
                                  rownames_to_column("gene") %>%
                                  mutate(celltype = current_cell)
  #assess significance
  volcanodata[[current_cell]]$DE <- "None"
  volcanodata[[current_cell]][which(volcanodata[[current_cell]]$p_val_adj < 0.05 & 
                                      volcanodata[[current_cell]]$avg_log2FC >= cutoff), "DE"] <- ident1 #positive avg_log2FC are associated with ident.1
  volcanodata[[current_cell]][which(volcanodata[[current_cell]]$p_val_adj < 0.05 & 
                                      volcanodata[[current_cell]]$avg_log2FC <= -cutoff), "DE"] <- ident2 #negative avg_log2FC are associated with ident.2
  allvolcano <- rbind(allvolcano, volcanodata[[current_cell]])
}

#label individual genes
delabels <- allvolcano %>% filter(DE != 'None') #only label significant genes
delabels <- delabels %>% filter(gene %in% goi) #filter based on genes of interest

allvolcano %>%
    ggplot(aes(x = avg_log2FC,
               y = -log10(p_val_adj))) + 
        geom_hline(yintercept = -log10(0.05),
                   size = 0.5,
                   color = "grey90",
                   alpha = 0.5) + 
        geom_vline(xintercept = c(-cutoff, cutoff),
                   size = 0.5,
                   color = "grey90",
                   alpha = 0.5) + 
        geom_point(aes(color = DE),
                   size = 3,
                   alpha = 0.5,
                   shape = 16) +
        scale_color_manual(values = c("None" = "grey70",
                                      custom_colors("treatments"))) +
        scale_y_continuous(limits = c(0,10)) +
        theme_minimal() +
        theme(panel.grid = element_blank(),
              text = element_text(size = 15),
              panel.border = element_rect(fill = NA, color = "black"),
              axis.text = element_text(), 
              panel.spacing = unit(0, "lines"),
              axis.title = element_text()) +
        labs(x = paste("Average Log2 Fold Change (",ident1,"/",ident2,")"),
             y = expression(-log[10]~(Adjusted~p~Value)),
             title="Pancreas-matching T cells in blood") +
        geom_label_repel(data = delabels,
                         aes(x = avg_log2FC,
                             y = -log10(p_val_adj),
                             label = gene,
                             color = DE),
                         label.size = NA,
                         label.padding = 0.2, 
                         na.rm = TRUE,
                         fill = alpha(c("white"), 0.8), 
                         size=7, 
                         max.overlaps = 50) +
  facet_wrap(~celltype)

## RECORD OVERLAPS OF GENES FOR VENN DIAGRAM
allvolcano %>% filter(celltype=="CD8" & DE == "PD1") %>% pull(gene) -> up_pd1_v_igg

ident1 <- "Spt"
ident2 <- "IgG"

Idents(seurat_integrated) <- seurat_integrated$tissue
seurat_bld <- subset(seurat_integrated, idents = "blood") #subset on T cells in blood
Idents(seurat_bld) <- seurat_bld$pancreas_matching
seurat_bld <- subset(seurat_bld, idents = "PM") #subset on PM T cells only
Idents(seurat_bld) <- seurat_bld$celltype

volcanodata <- list()
allvolcano <- data.frame()
for (current_cell in c("CD8", "Tcon", "Treg")) {
  volcanodata[[current_cell]] <- FindMarkers(seurat_bld, #define the table of DE genes
                                     ident.1 = ident1,
                                     ident.2 = ident2, 
                                     verbose = F,
                                     group.by = "treatment",
                                     subset.ident = current_cell,
                                     only.pos = F, 
                                     logfc.threshold = cutoff, 
                                     assay = "RNA") %>% 
                                  rownames_to_column("gene") %>%
                                  mutate(celltype = current_cell)
  #assess significance
  volcanodata[[current_cell]]$DE <- "None"
  volcanodata[[current_cell]][which(volcanodata[[current_cell]]$p_val_adj < 0.05 & 
                                      volcanodata[[current_cell]]$avg_log2FC >= cutoff), "DE"] <- ident1 #positive avg_log2FC are associated with ident.1
  volcanodata[[current_cell]][which(volcanodata[[current_cell]]$p_val_adj < 0.05 & 
                                      volcanodata[[current_cell]]$avg_log2FC <= -cutoff), "DE"] <- ident2 #negative avg_log2FC are associated with ident.2
  allvolcano <- rbind(allvolcano, volcanodata[[current_cell]])
}

## RECORD OVERLAPS OF GENES FOR VENN DIAGRAM
allvolcano %>% filter(celltype=="CD8" & DE == "Spt") %>% pull(gene) -> up_spt_v_igg

intersect(up_pd1_v_igg, up_spt_v_igg) #there is no overlap

```

DE genes volcano plot between pancreas-matching and non-matching T cells in the blood between different treatment groups.
```{r volcano plot of PM vs. non-PM T cells in blood}
cutoff <- 0.25 #set a cutoff for DE gene expression
ident1 <- "PM"
ident2 <- "non-PM"

Idents(seurat_integrated) <- seurat_integrated$tissue
seurat_bld <- subset(seurat_integrated, idents = "blood") #subset on T cells in blood
Idents(seurat_bld) <- seurat_bld$treatment

allvolcano <- data.frame()
for (current_treatment in c("IgG","Spt","PD1")) {
  seurat_tmp <- subset(seurat_bld, idents = current_treatment) #subset on PM T cells only
  Idents(seurat_tmp) <- seurat_tmp$celltype
  
  volcanodata <- list()
  for (current_cell in c("CD8", "Tcon", "Treg")) {
    volcanodata[[current_cell]] <- FindMarkers(seurat_tmp, #define the table of DE genes
                                       ident.1 = ident1,
                                       ident.2 = ident2, 
                                       verbose = F,
                                       group.by = "pancreas_matching",
                                       subset.ident = current_cell,
                                       only.pos = F, 
                                       logfc.threshold = cutoff, 
                                       assay = "RNA") %>% 
                                    rownames_to_column("gene") %>%
                                    mutate(celltype = current_cell,
                                           treatment = current_treatment)
    #assess significance
    volcanodata[[current_cell]]$DE <- "None"
    volcanodata[[current_cell]][which(volcanodata[[current_cell]]$p_val_adj < 0.05 & 
                                        volcanodata[[current_cell]]$avg_log2FC >= cutoff), "DE"] <- ident1 #positive avg_log2FC are associated with ident.1
    volcanodata[[current_cell]][which(volcanodata[[current_cell]]$p_val_adj < 0.05 & 
                                        volcanodata[[current_cell]]$avg_log2FC <= -cutoff), "DE"] <- ident2 #negative avg_log2FC are associated with ident.2
    allvolcano <- rbind(allvolcano, volcanodata[[current_cell]])
  }
}
#label individual genes
delabels <- allvolcano %>% filter(DE != 'None') #only label significant genes
delabels <- delabels %>% filter(gene %in% goi) #filter based on genes of interest

allvolcano %>%
    ggplot(aes(x = avg_log2FC,
               y = -log10(p_val_adj))) + 
        geom_hline(yintercept = -log10(0.05),
                   size = 0.5,
                   color = "grey90",
                   alpha = 0.5) + 
        geom_vline(xintercept = c(-cutoff, cutoff),
                   size = 0.5,
                   color = "grey90",
                   alpha = 0.5) + 
        geom_point(aes(color = DE),
                   size = 3,
                   alpha = 0.5,
                   shape = 16) +
        scale_color_manual(values = c("None" = "grey70",
                                      custom_colors("pm"))) +
        scale_y_continuous() +
        theme_minimal() +
        theme(panel.grid = element_blank(),
              text = element_text(size = 15),
              panel.border = element_rect(fill = NA, color = "black"),
              axis.text = element_text(), 
              panel.spacing = unit(0, "lines"),
              axis.title = element_text()) +
        labs(x = paste("Average Log2 Fold Change (",ident1,"/",ident2,")"),
             y = expression(-log[10]~(Adjusted~p~Value)),
             title="T cells in blood") +
        geom_label_repel(data = delabels,
                         aes(x = avg_log2FC,
                             y = -log10(p_val_adj),
                             label = gene,
                             color = DE),
                         label.size = NA,
                         label.padding = 0.2, 
                         na.rm = TRUE,
                         fill = alpha(c("white"), 0.8), 
                         size = 5, 
                         max.overlaps = 50) +
  facet_wrap(celltype ~ treatment)

## RECORD OVERLAPS OF GENES FOR VENN DIAGRAM
up_pm <- list()
allvolcano %>% filter(celltype == "CD8" & DE == "PM" & treatment == "IgG") %>% pull(gene) -> up_pm[["IgG"]]
allvolcano %>% filter(celltype == "CD8" & DE == "PM" & treatment == "PD1") %>% pull(gene) -> up_pm[["PD1"]]
allvolcano %>% filter(celltype == "CD8" & DE == "PM" & treatment == "Spt") %>% pull(gene) -> up_pm[["Spt"]]

length(setdiff(up_pm[["IgG"]],c(up_pm[["PD1"]], up_pm[["Spt"]])))
length(setdiff(up_pm[["Spt"]],c(up_pm[["PD1"]], up_pm[["IgG"]])))
length(setdiff(up_pm[["PD1"]],c(up_pm[["Spt"]], up_pm[["IgG"]])))

common_to_all_treatments <- Reduce(intersect, list(up_pm[["PD1"]],up_pm[["Spt"]],up_pm[["IgG"]]))
length(common_to_all_treatments)

length(setdiff(Reduce(intersect, list(up_pm[["PD1"]], up_pm[["Spt"]])), common_to_all_treatments))
length(setdiff(Reduce(intersect, list(up_pm[["PD1"]], up_pm[["IgG"]])), common_to_all_treatments))
length(setdiff(Reduce(intersect, list(up_pm[["IgG"]], up_pm[["Spt"]])), common_to_all_treatments))

write.csv(rbind(data.frame(gene = setdiff(up_pm[["IgG"]],c(up_pm[["PD1"]], up_pm[["Spt"]])), treatment_group = "IgG"),
      data.frame(gene = setdiff(up_pm[["Spt"]],c(up_pm[["PD1"]], up_pm[["IgG"]])), treatment_group = "Spt"),
      data.frame(gene = setdiff(up_pm[["PD1"]],c(up_pm[["IgG"]], up_pm[["Spt"]])), treatment_group = "PD1"),
      data.frame(gene = common_to_all_treatments, treatment_group = "IgG-Spt-PD1"),
      data.frame(gene = setdiff(Reduce(intersect, list(up_pm[["PD1"]], up_pm[["Spt"]])), common_to_all_treatments), treatment_group = "PD1-Spt"),
      data.frame(gene = setdiff(Reduce(intersect, list(up_pm[["PD1"]], up_pm[["IgG"]])), common_to_all_treatments), treatment_group = "PD1-IgG"),
      data.frame(gene = setdiff(Reduce(intersect, list(up_pm[["IgG"]], up_pm[["Spt"]])), common_to_all_treatments), treatment_group = "IgG-Spt")),
      "output/upregulated_genes.tsv", 
      quote = F, 
      col.names = F, 
      row.names = F)

dn_pm <- list()
allvolcano %>% filter(celltype == "CD8" & DE == "non-PM" & treatment == "IgG") %>% pull(gene) -> dn_pm[["IgG"]]
allvolcano %>% filter(celltype == "CD8" & DE == "non-PM" & treatment == "PD1") %>% pull(gene) -> dn_pm[["PD1"]]
allvolcano %>% filter(celltype == "CD8" & DE == "non-PM" & treatment == "Spt") %>% pull(gene) -> dn_pm[["Spt"]]

length(setdiff(dn_pm[["IgG"]],c(dn_pm[["PD1"]], dn_pm[["Spt"]])))
length(setdiff(dn_pm[["Spt"]],c(dn_pm[["PD1"]], dn_pm[["IgG"]])))
length(setdiff(dn_pm[["PD1"]],c(dn_pm[["Spt"]], dn_pm[["IgG"]])))

common_to_all_treatments <- Reduce(intersect, list(dn_pm[["PD1"]],dn_pm[["Spt"]],dn_pm[["IgG"]]))
length(common_to_all_treatments)

length(setdiff(Reduce(intersect, list(dn_pm[["PD1"]], dn_pm[["Spt"]])), common_to_all_treatments))
length(setdiff(Reduce(intersect, list(dn_pm[["PD1"]], dn_pm[["IgG"]])), common_to_all_treatments))
length(setdiff(Reduce(intersect, list(dn_pm[["IgG"]], dn_pm[["Spt"]])), common_to_all_treatments))

write.csv(rbind(data.frame(gene = setdiff(dn_pm[["IgG"]],c(dn_pm[["PD1"]], dn_pm[["Spt"]])), treatment_grodn = "IgG"),
      data.frame(gene = setdiff(dn_pm[["Spt"]],c(dn_pm[["PD1"]], dn_pm[["IgG"]])), treatment_grodn = "Spt"),
      data.frame(gene = setdiff(dn_pm[["PD1"]],c(dn_pm[["IgG"]], dn_pm[["Spt"]])), treatment_grodn = "PD1"),
      data.frame(gene = common_to_all_treatments, treatment_grodn = "IgG-Spt-PD1"),
      data.frame(gene = setdiff(Reduce(intersect, list(dn_pm[["PD1"]], dn_pm[["Spt"]])), common_to_all_treatments), treatment_grodn = "PD1-Spt"),
      data.frame(gene = setdiff(Reduce(intersect, list(dn_pm[["PD1"]], dn_pm[["IgG"]])), common_to_all_treatments), treatment_grodn = "PD1-IgG"),
      data.frame(gene = setdiff(Reduce(intersect, list(dn_pm[["IgG"]], dn_pm[["Spt"]])), common_to_all_treatments), treatment_grodn = "IgG-Spt")),
      "output/dnregulated_genes.tsv", 
      quote = F, 
      col.names = F, 
      row.names = F)

#BARCHARTS OF DE GENES ACROSS PM AND NON-PM
Idents(seurat_bld) <- seurat_bld$celltype
seurat_bld_cd8 <- subset(seurat_bld, idents = "CD8")
Idents(seurat_bld_cd8) <- seurat_bld_cd8$pancreas_matching

#Genes different in PM vs. non-PM across all treatment groups
newdata <- FindMarkers(seurat_bld_cd8, #define the table of DE genes
                       ident.1 = "PM",
                       ident.2 = "non-PM",
                       verbose = F,
                       group.by = "pancreas_matching",
                       only.pos = F, 
                       logfc.threshold = cutoff, 
                       assay = "RNA") %>% 
  rownames_to_column("gene") %>%
  filter(p_val_adj < 0.05 & p_val < 0.01 & abs(avg_log2FC) > 1) %>% 
  arrange(desc(avg_log2FC))
#assess significance
newdata$DE <- "None"
newdata[which(newdata$avg_log2FC >= cutoff), "DE"] <- "PM" 
newdata[which(newdata$avg_log2FC <= -cutoff), "DE"] <- "non-PM" 
newdata$gene <- factor(newdata$gene, levels = rev(newdata$gene))
#plot
newdata %>% ggplot(aes(y = gene, x = avg_log2FC, fill = DE, group = gene)) + 
  geom_bar(stat="identity") + 
  theme_minimal() +
  theme(panel.grid = element_blank(),
        text = element_text(size = 15),
        panel.border = element_rect(fill = NA, color = "black"),
        axis.text = element_text(), 
        axis.title.y = element_blank()) +
  scale_fill_manual(values = custom_colors("pm")) +
  geom_text(aes(label = signif(p_val_adj, digits = 3)))

#Genes different in PD1 versus others within PM cells
Idents(seurat_bld_cd8) <- seurat_bld_cd8$pancreas_matching
seurat_bld_cd8_pm <- subset(seurat_bld_cd8, idents = "PM")
newdata <- FindMarkers(seurat_bld_cd8_pm, #define the table of DE genes
                       ident.1 = "PD1",
                       verbose = F,
                       group.by = "treatment",
                       only.pos = F, 
                       logfc.threshold = cutoff, 
                       assay = "RNA") %>% 
  rownames_to_column("gene") %>%
  filter(p_val_adj < 0.05 & p_val < 0.01 & abs(avg_log2FC) > 0.57) %>% 
  arrange(desc(avg_log2FC))
#assess significance
newdata$DE <- "None"
newdata[which(newdata$avg_log2FC >= cutoff), "DE"] <- "PD1" 
newdata[which(newdata$avg_log2FC <= -cutoff), "DE"] <- "other" 
newdata$gene <- factor(newdata$gene, levels = rev(newdata$gene))
#plot
newdata %>% ggplot(aes(y = gene, x = avg_log2FC, fill = DE)) + 
  geom_bar(stat="identity") + 
  theme_minimal() +
  theme(panel.grid = element_blank(),
        text = element_text(size = 15),
        panel.border = element_rect(fill = NA, color = "black"),
        axis.text = element_text(), 
        axis.title.y = element_blank()) +
  scale_fill_manual(values = custom_colors("treatments")) +
  geom_text(aes(label = signif(p_val_adj, digits = 3)))

#get genes different in PD1 PM versus all else (both non-PM and PM)
new_metadata <- seurat_bld_cd8@meta.data %>% 
                  mutate(treatment_pancreas_matching = ifelse(is.na(clonotype_id), NA_character_, paste0(treatment, "_", pancreas_matching))) %>% 
                  dplyr::select(treatment_pancreas_matching)
seurat_bld_cd8 <- AddMetaData(seurat_bld_cd8, new_metadata)

newdata <- FindMarkers(seurat_bld_cd8, #define the table of DE genes
                       ident.1 = "PD1_PM",
                       verbose = F,
                       group.by = "treatment_pancreas_matching",
                       only.pos = F, 
                       logfc.threshold = cutoff, 
                       assay = "RNA") %>% 
  rownames_to_column("gene") %>%
  filter(p_val_adj < 0.05 & p_val < 0.01 & abs(avg_log2FC) > 1) %>% 
  arrange(desc(avg_log2FC))
#assess significance
newdata$DE <- "None"
newdata[which(newdata$avg_log2FC >= cutoff), "DE"] <- "PM" 
newdata[which(newdata$avg_log2FC <= -cutoff), "DE"] <- "non-PM" 
newdata$gene <- factor(newdata$gene, levels = rev(newdata$gene))
#plot
newdata %>% ggplot(aes(y = gene, x = avg_log2FC, fill = DE)) + 
  geom_bar(stat="identity") + 
  theme_minimal() +
  theme(panel.grid = element_blank(),
        text = element_text(size = 15),
        panel.border = element_rect(fill = NA, color = "black"),
        axis.text = element_text(), 
        axis.title.y = element_blank()) +
  scale_fill_manual(values = custom_colors("pm")) +
  geom_text(aes(label = signif(p_val_adj, digits = 3)))
```

```{r representative UMAPs of PM and non-PM clusters}

seurat_integrated@meta.data %>% 
  filter(!is.na(pancreas_matching)) %>%
  dplyr::select(treatment,
                pancreas_matching, 
                seurat_clusters,
                allcells_UMAP_1, 
                allcells_UMAP_2) %>% 
  group_by(seurat_clusters) %>% 
  mutate(U1 = mean(allcells_UMAP_1),
         U2 = mean(allcells_UMAP_2)) %>%
  group_by(treatment,
           pancreas_matching, 
           seurat_clusters, U1, U2) %>% 
  summarise(num_cells = n(),
            .groups = "keep") %>%
  ggplot() +
  geom_point(data = seurat_integrated@meta.data %>%
               dplyr::select(-treatment) %>%
               dplyr::select(-pancreas_matching),
             aes(x = allcells_UMAP_1, 
                 y = allcells_UMAP_2),
             shape = 16,
             color = "grey90") +
  geom_point(aes(x = U1,
                 y = U2,
                 size = num_cells,
                 color = seurat_clusters),
             shape = 16) +
  theme_minimal() +
  theme(axis.title = element_blank(),
        axis.text = element_blank(),
        panel.grid = element_blank(),
        legend.position = "none",
        text = element_blank()) +
  labs(color = NULL) +
  scale_color_manual(values = custom_colors("clusters")) +
  scale_size_continuous(range = c(5,25)) +
  guides(color = guide_legend(override.aes = list(alpha = 1, 
                                                  size = 5, 
                                                  shape = 19))) +
  facet_wrap(treatment ~ pancreas_matching, ncol = 2) +
  geom_text(aes(x = U1, y = U2, label = num_cells), color = "black", size = 5, fontface = "bold")
```

Plot how clones that "feed" into the main clusters in the pancreas differ in transcriptional profile. Stacked bar plots and UMAPs will be generated to highlight the differences.
```{r blood clusters that contribute to pancreas clusters across treatments}
clones <- list() #list of clonotypes within a particular pancreas-associated cluster of interest
blood_feeder_clusters <- data.frame() #data.frame for all of the data of each cluster in the blood "feeding" into the cluster of interest
for (current_cluster in c("Recently_activated","CD8_eff","CD8_texh")) {
  for (current_treatment in c("IgG","Spt","PD1")) {
    clones[[current_treatment]] <- seurat_integrated@meta.data %>% 
                                  filter(tissue == "pancreas" & 
                                           clonotype_tissue_groups %in% c("pLN&blood&pancreas","blood&pancreas") & 
                                           treatment == current_treatment & 
                                           seurat_clusters == current_cluster) %>% 
                                  pull(clonotype_id) %>%
                                  unique()
    blood_feeder_clusters <- rbind(blood_feeder_clusters, seurat_integrated@meta.data %>% 
                                    filter(tissue == "blood" & clonotype_id %in% clones[[current_treatment]]) %>%
                                    group_by(seurat_clusters) %>%
                                    tally(name = "num_cells") %>%
                                    mutate(treatment = current_treatment,
                                           pan_seurat_cluster = current_cluster))
  }
}
blood_feeder_clusters$treatment <- factor(blood_feeder_clusters$treatment, levels = c("IgG","Spt","PD1"))
#blood_feeder_clusters$pan_seurat_cluster <- factor(blood_feeder_clusters$pan_seurat_cluster, levels = c("Recently_activated","CD8_eff","CD8_texh"))

blood_feeder_clusters %>% 
  ggplot(aes(x = treatment, 
             y = num_cells,
             fill = seurat_clusters)) + 
  geom_bar(position = "fill", stat = "identity") +
  scale_fill_manual(values = custom_colors("clusters")) +
  scale_y_continuous(expand = c(0,0)) +
  scale_x_discrete(expand = c(0,0)) +
  theme_minimal() +
  theme(panel.grid = element_blank(),
        text = element_text(size = 10),
        axis.title.x = element_blank(),
        #panel.border = element_rect(color = "black", fill = NA, size = 1),
        legend.position = "none") +
  labs(y = expression("Percent of cells of clonotype in blood")) +
  facet_wrap(~ pan_seurat_cluster)

##TCON

clones <- list() #list of clonotypes within a particular pancreas-associated cluster of interest
blood_feeder_clusters <- data.frame() #data.frame for all of the data of each cluster in the blood "feeding" into the cluster of interest
for (current_cluster in c("Recently_activated","Tcon_eff","Tcon_effmem","Tcon_Tfh")) {
  for (current_treatment in c("IgG","Spt","PD1")) {
    clones[[current_treatment]] <- seurat_integrated@meta.data %>% 
                                  filter(tissue == "pancreas" & 
                                           clonotype_tissue_groups %in% c("pLN&blood&pancreas","blood&pancreas") & 
                                           treatment == current_treatment & 
                                           seurat_clusters == current_cluster) %>% 
                                  pull(clonotype_id) %>%
                                  unique()
    blood_feeder_clusters <- rbind(blood_feeder_clusters, seurat_integrated@meta.data %>% 
                                    filter(tissue == "blood" & clonotype_id %in% clones[[current_treatment]]) %>%
                                    group_by(seurat_clusters) %>%
                                    tally(name = "num_cells") %>%
                                    mutate(treatment = current_treatment,
                                           pan_seurat_cluster = current_cluster))
  }
}
blood_feeder_clusters$treatment <- factor(blood_feeder_clusters$treatment, levels = c("IgG","Spt","PD1"))
#blood_feeder_clusters$pan_seurat_cluster <- factor(blood_feeder_clusters$pan_seurat_cluster, levels = c("Recently_activated","CD8_eff","CD8_texh"))

blood_feeder_clusters %>% 
  ggplot(aes(x = treatment, 
             y = num_cells,
             fill = seurat_clusters)) + 
  geom_bar(position = "fill", stat = "identity") +
  scale_fill_manual(values = custom_colors("clusters")) +
  scale_y_continuous(expand = c(0,0)) +
  scale_x_discrete(expand = c(0,0)) +
  theme_minimal() +
  theme(panel.grid = element_blank(),
        text = element_text(size = 10),
        axis.title.x = element_blank(),
        #panel.border = element_rect(color = "black", fill = NA, size = 1),
        legend.position = "none") +
  labs(y = expression("Percent of cells of clonotype in blood")) +
  facet_wrap(~ pan_seurat_cluster)
```

SAVE data for COMETSC
```{r COMET}
bld <- subset(seurat_integrated,idents = "blood")
Idents(bld) <- bld$cd4_or_cd8
bld8 <- subset(bld, idents = "CD8")
bld8 <- subset(bld8, cells = bld@meta.data %>% filter(!is.na(pancreas_matching)) %>% rownames())
write.table(bld8@meta.data %>% mutate(comet_clusters = paste0(treatment,"_",pancreas_matching))%>%dplyr::select(comet_clusters), "output/cometsc/cluster.txt", quote = F, sep = "\t", col.names = F)
write.table(bld8@meta.data %>% dplyr::select(allcells_UMAP_1,allcells_UMAP_2), "output/cometsc/vis.txt", quote = F, sep = "\t", col.names = F)
write.table(as.data.frame(bld8@assays$RNA@data), "output/cometsc/marker.txt", quote = F, sep = "\t")

```

Upset plot for sharing between tissues

