---
title: "7_figure6_plots"
output: rmarkdown::github_document
date: '2022-07-10'
---


This code documents the initialization of the Seurat objects for patient colitis.

```{r setup, include=FALSE, echo=FALSE}
knitr::opts_chunk$set(echo = TRUE, root.dir = "C:/Users/Strix/GitRepos/NOD_spt_pd1")
knitr::opts_knit$set(echo = TRUE, root.dir = "C:/Users/Strix/GitRepos/NOD_spt_pd1")
```

```{r libraries, message = FALSE, warning = FALSE}
require(dplyr, quietly = T) #simple manipulation of data.frames
require(Seurat, quietly = T) #scRNA-seq analysis
require(reshape, quietly = T) #melting and casting data.frames
require(ggplot2, quietly = T) #plotting
require(rstatix, quietly = T) #adding p values to plots
require(ggprism, quietly = T) #adding p values to plots
require(ggsci, quietly = T) #IGV color palette
require(patchwork, quietly = T) #adding p values to plots
require(magrittr, quietly = T) #adding p values to plots
require(gridExtra, quietly = T) #display multiple plots
require(ggrepel, quietly = T) #labels for plots
require(tibble, quietly = T) #rownames_to_column function
require(useful, quietly = T) #corner function to peek at data.frames
require(viridis, quietly = T) #viridis color palettes
require(readxl, quietly = T) #reading Excel files
require(msigdbr, quietly = T) #accessing gene signatures from mSigDB
source("scripts/utils.R") #custom color palettes and other functions
set.seed(1) #set a seed for reproducibility
```

## Initialize Seurat object

Matrices, genes, barcodes, filtered TCR contig annotations, and consensus TCR annotations for each sample are within the data/ directory named as **GSE180045**. Initiate the Seurat object by reading in each individual output from cellranger multi using a loop and subsequent merge. Loop through each sample in samples 1) storing Gene Expression (GEX) and feature-barcoding (FB) data; 2) creating list of seurat objects with GEX and FB assay data under the list name of each sample; 3) sample meta data (tissue: pancreas [pan], pancreatic lymph node [pLN], blood [bld] and antibody treatment [isotype: IgG or anti-PD-1: PD1] or spontaneously diabetic mice [Spt]).
```{r PBMC seurat initialization}
conditions <- list.files("data/GSE180045") #all samples are within individual folders
conditions

data <- list() #list containing GEX and FB data
seurat_objects <- list() #list containing all the Seurat objects

for (condition in conditions) {
  data[[condition]] <- Read10X(data.dir = paste0("data/GSE180045/",condition), strip.suffix = T)  #read 10X data
  #create Seurat object with gex data
  seurat_objects[[condition]] <- CreateSeuratObject(counts = data[[condition]][['Gene Expression']], 
                                                    project = condition, min.cells = 50, min.features = 200) 
  
  #add FB assay data to object with columns for only the cells (colnames) in original GEX data
  seurat_objects[[condition]][['FB']] <- CreateAssayObject(counts = 
                                        data[[condition]][['Antibody Capture']][,colnames(seurat_objects[[condition]])])
  
  ##add tissue and treatment meta data to Seurat object
  seurat_objects[[condition]]@meta.data$timepoint <- ifelse(strsplit(condition,"_")[[1]][2] == "L", "late", "early")
  seurat_objects[[condition]]@meta.data$adverse_event <- case_when(substr(seurat_objects[[condition]]$orig.ident,1,1) == "A" ~ "none",
                                                      substr(seurat_objects[[condition]]$orig.ident,1,1) %in% c("B","C") ~ "irAE")
}

# Merge all Seurat objects generating unique cell IDs
seurat_object <- merge(x = seurat_objects[[conditions[1]]], 
                       y = seurat_objects[2:length(seurat_objects)], 
                       add.cell.ids = conditions)

seurat_object$adverse_event <- factor(seurat_object$adverse_event, levels = c("none","irAE"))

#the Seurat data has already been filtered
seurat_object[["percent.mt"]] <- PercentageFeatureSet(seurat_object, pattern = "^mt-", assay = "RNA") #percent mitochondrial RNA

#cluster the data 
seurat_object <- ScaleData(seurat_object, verbose = FALSE) #scale the RNA data
seurat_object <- RunPCA(seurat_object, npcs = 30, verbose = FALSE, features = rownames(x = seurat_object)) #run PCA
seurat_object <- RunUMAP(seurat_object, reduction = "pca", dims = 1:30)
seurat_object <- FindNeighbors(seurat_object, reduction = "pca", dims = 1:30)
seurat_object <- FindClusters(seurat_object, resolution = 0.2)
p_cluster <- DimPlot(seurat_object, reduction = "umap", label = TRUE, label.size = 5) + theme(legend.position = "none")
p_markers <- FeaturePlot(seurat_object, "CD3E")
p_cluster + p_markers

seurat_markers <- FindAllMarkers(seurat_integrated, min.pct = 0.25, only.pos = T, logfc.threshold = 0.25) %>%
                     group_by(cluster) %>%
                     filter(p_val_adj < 0.05) %>%
                     group_by(cluster) %>%
                     arrange(desc(avg_log2FC), .by_group = T) %>%
                     top_n(n = 50, wt = avg_log2FC)

tcells <- c(WhichCells(pbmc, idents = c(2,5,7,8,10,11,15)), WhichCells(pbmc, expression = CD3E > 0))
seurat_object <- subset(seurat_object, cells = tcells)

seurat_object <- ScaleData(seurat_object, verbose = FALSE) #scale the RNA data
seurat_object <- RunPCA(seurat_object, npcs = 30, verbose = FALSE, features = rownames(x = seurat_object)) #run PCA
seurat_object <- RunUMAP(seurat_object, reduction = "pca", dims = 1:30)
seurat_object <- FindNeighbors(seurat_object, reduction = "pca", dims = 1:30)
seurat_object <- FindClusters(seurat_object, resolution = 0.2)
p_cluster <- DimPlot(seurat_object, reduction = "umap", label = TRUE, label.size = 5) + theme(legend.position = "none")
p_markers <- FeaturePlot(seurat_object, "CD8A", max.cutoff = 10)
p_cluster + p_markers

saveRDS(seurat_object, "objects/pbmc/seurat_pbmc.rds")
```

```{r Seurat integration by tissue}
split_seurat <- SplitObject(seurat_object, split.by = "orig.ident") # Split the seurat object for finding integration anchors

# Normalize the split data by tissue and define variable features for each
split_seurat <- lapply(X = split_seurat, FUN = function(x) {
  x <- NormalizeData(x)
  x <- FindVariableFeatures(x, selection.method = "vst", nfeatures = 2000)
})

seurat_anchors <- FindIntegrationAnchors(object.list = split_seurat, dims = 1:20) # Define anchors and integrate data (warning: takes a long time)
saveRDS(seurat_anchors, file='objects/pbmc/pbmc_anchors.rds') # Save anchors just in case

seurat_integrated <- IntegrateData(anchorset = seurat_anchors, dims = 1:20) # Integrate Seurat data using anchors
DefaultAssay(seurat_integrated) <- "integrated"

saveRDS(seurat_integrated, file='objects/pbmc/pbmc_integrated.rds') # Save integrated Seurat object
```

```{r cluster}

new_metadata <- data.frame(coinh_count = rowSums(as.data.frame(t(as.data.frame(seurat_integrated@assays$RNA@data[c("PDCD1","HAVCR2","CTLA4","LAG3","TIGIT","CD160"),]))) > 0))
seurat_integrated <- AddMetaData(seurat_integrated, new_metadata)


seurat_integrated <- ScaleData(seurat_integrated, verbose = FALSE) #scale the RNA data
seurat_integrated <- RunPCA(seurat_integrated, npcs = 30, verbose = FALSE, features = rownames(x = seurat_integrated)) #run PCA
seurat_integrated <- RunUMAP(seurat_integrated, reduction = "pca", dims = 1:30)
seurat_integrated <- FindNeighbors(seurat_integrated, reduction = "pca", dims = 1:30)
seurat_integrated <- FindClusters(seurat_integrated, resolution = 0.9)
p_cluster <- DimPlot(seurat_integrated, reduction = "umap", label = TRUE, label.size = 5) + theme(legend.position = "none")
p_markers <- DimPlot(seurat_integrated, group.by = "adverse_event")
p_cluster + p_markers
```
```{r cluster}
new_metadata <- data.frame(t(as.matrix(seurat_integrated@assays$RNA[c("CD8A", "CD4","KLRK1","ITGA4","CXCR3","ITGB1","TIGIT","CD38"),])))
seurat_integrated <- AddMetaData(seurat_integrated, new_metadata)

seurat_integrated@meta.data %>% 
    group_by(seurat_clusters) %>% 
    summarize(CD8A = mean(CD8A), CD4 = mean(CD4)) %>% 
    mutate(ratio = CD8A/CD4) %>%
    filter(ratio > 1) %>% 
    pull(seurat_clusters)

cd8_clusters <- c(1,19)

seurat_integrated <- subset(seurat_integrated, idents = cd8_clusters)
seurat_integrated <- ScaleData(seurat_integrated, verbose = FALSE) #scale the RNA data
seurat_integrated <- RunPCA(seurat_integrated, npcs = 30, verbose = FALSE, features = rownames(x = seurat_integrated)) #run PCA
seurat_integrated <- RunUMAP(seurat_integrated, reduction = "pca", dims = 1:30)
seurat_integrated <- FindNeighbors(seurat_integrated, reduction = "pca", dims = 1:30)
seurat_integrated <- FindClusters(seurat_integrated, resolution = 0.8)
p_cluster <- DimPlot(seurat_integrated, reduction = "umap", label = TRUE, label.size = 5) + theme(legend.position = "none")
p_markers <- DimPlot(seurat_integrated, group.by = "adverse_event")
p_cluster + p_markers

saveRDS(seurat_integrated, "objects/pbmc/seurat_CD8.rds")
saveRDS(cd8_markers, "objects/pbmc/seurat_CD8_markers.rds")
```


```{r project gene signatures onto PBMC}

ifn <- FindMarkers(seurat_integrated, 
                       ident.1 = "Interferon_sensing",
                      group.by = "seurat_clusters") %>% 
            filter(p_val_adj < 0.05) %>%
            arrange(desc(avg_log2FC))
texh <- FindMarkers(seurat_integrated, 
                       ident.1 = "CD8_texh",
                      group.by = "seurat_clusters") %>% 
            filter(p_val_adj < 0.05) %>%
            arrange(desc(avg_log2FC))
prolif <- FindMarkers(seurat_integrated, 
                       ident.1 = "Proliferating",
                      group.by = "seurat_clusters") %>% 
            filter(p_val_adj < 0.05) %>%
            arrange(desc(avg_log2FC))
eff <- FindMarkers(seurat_integrated, 
                       ident.1 = "CD8_eff",
                      ident.2 = "CD8_cm",
                      group.by = "seurat_clusters") %>% 
            filter(p_val_adj < 0.05) %>%
            arrange(desc(avg_log2FC))
pm <- FindMarkers(seurat_integrated, 
                       ident.1 = "PM",
                      ident.2 = "non-PM",
                      group.by = "pancreas_matching") %>% 
            filter(p_val_adj < 0.05) %>%
            arrange(desc(avg_log2FC))

ifn_signature <- convertMouseGeneList(ifn %>% filter(avg_log2FC > 0) %>% rownames()) #convert the positive logFC genes to human
texh_signature <- convertMouseGeneList(texh %>% filter(avg_log2FC > 0) %>% rownames()) #convert the positive logFC genes to human
prolif_signature <- convertMouseGeneList(prolif %>% filter(avg_log2FC > 0) %>% rownames()) #convert the positive logFC genes to human
eff_signature <- convertMouseGeneList(eff %>% filter(avg_log2FC > 0) %>% rownames()) #convert the positive logFC genes to human
pm_signature <- convertMouseGeneList(pm %>% filter(avg_log2FC > 0) %>% rownames()) #convert the positive logFC genes to human

cd8 <- AddModuleScore(cd8, features = list(ifn_signature), name = "ifn_signature")
cd8 <- AddModuleScore(cd8, features = list(texh_signature), name = "texh_signature")
cd8 <- AddModuleScore(cd8, features = list(prolif_signature), name = "prolif_signature")
cd8 <- AddModuleScore(cd8, features = list(eff_signature), name = "eff_signature")
cd8 <- AddModuleScore(cd8, features = list(pm_signature), name = "pm_signature")

Tcon_prog <- FindMarkers(seurat_integrated, 
                       ident.1 = "Tcon_prog",
                      group.by = "seurat_clusters") %>% 
            filter(p_val_adj < 0.05) %>%
            arrange(desc(avg_log2FC))

Tcon_prog_signature <- convertMouseGeneList(Tcon_prog %>% filter(avg_log2FC > 0) %>% rownames()) #convert the positive logFC genes to human

cd4 <- AddModuleScore(cd4, features = list(Tcon_prog_signature), name = "Tcon_prog_signature")

```

```{r plot the data}
stat.test <- data.frame(cd8@meta.data %>%
                            group_by(adverse_event,orig.ident) %>%
                            summarize(ifn_signature1 = mean(texh_signature1)) %>%
                            ungroup() %>%
                            t_test(ifn_signature1 ~ adverse_event) %>% 
                            add_significance() %>%
                            add_xy_position())

p_texh_patient <- cd8@meta.data %>%
            group_by(adverse_event, orig.ident) %>%
            summarize(texh_signature = mean(texh_signature1)) %>%
            ggplot(aes(x = adverse_event,
                       y = texh_signature)) +
            geom_violin(aes(
                       color = adverse_event,
                       fill = adverse_event),
                       scale = "width",
                        adjust = 1,
                        trim = F,
                        alpha = 0.5,
                        draw_quantiles = c(0.25,0.5,0.75),
                        lwd = 1) +
            geom_point(aes(color = adverse_event,
                       fill = adverse_event),
                       size = 5,
                       position = position_jitter(width = 0.4)) +
            theme_minimal() +
            theme(panel.grid = element_blank(),
                  panel.border = element_rect(color = "black", 
                                              fill = NA),
                  text = element_text(size = 15),
                  legend.position = "right") +
            scale_color_manual(values = custom_colors("colitis")) +
            scale_fill_manual(values = custom_colors("colitis")) +
            labs(title = "CD8+ T cells blood", 
                 y = "Terminally exhausted/effector-like gene signature", 
                 x = NULL,
                 color = NULL,
                 fill = NULL) +
            add_pvalue(stat.test,
                     label = "p.signif",
                     tip.length = 0,
                     label.size = 8)

stat.test <- data.frame(cd8@meta.data %>%
                            t_test(texh_signature1 ~ adverse_event) %>% 
                            add_significance() %>%
                            add_xy_position())

p_texh_cells <- cd8@meta.data %>%
            ggplot(aes(x = adverse_event,
                       y = texh_signature1)) +
            geom_violin(aes(
                       color = adverse_event,
                       fill = adverse_event),
                       scale = "width",
                        adjust = 1,
                        trim = F,
                        alpha = 0.5,
                        draw_quantiles = c(0.25,0.5,0.75),
                        lwd = 1) +
            theme_minimal() +
            theme(panel.grid = element_blank(),
                  panel.border = element_rect(color = "black", 
                                              fill = NA),
                  text = element_text(size = 15),
                  legend.position = "right") +
            scale_color_manual(values = custom_colors("colitis")) +
            scale_fill_manual(values = custom_colors("colitis")) +
            labs(title = "CD8+ T cells blood", 
                 y = "PD1 gene signature", 
                 x = NULL,
                 color = NULL,
                 fill = NULL) +
            add_pvalue(stat.test,
                     label = "p.signif",
                     tip.length = 0,
                     label.size = 8)

grid.arrange(p_texh_cells, p_texh_patient, ncol = 2)

new_metadata <- cd8@meta.data %>% mutate(texh_group = ifelse(texh_signature1 > mean(cd8$texh_signature1), "pos", "neg")) %>% dplyr::select(texh_group)
cd8 <- AddMetaData(cd8, new_metadata)

new_metadata <- data.frame(t(as.matrix(cd8@assays$RNA[c("CD38"),])))
cd8 <- AddMetaData(cd8, new_metadata)

new_metadata <- as.data.frame(cd8@reductions$umap@cell.embeddings)
cd8 <- AddMetaData(cd8, new_metadata)

cd8$adverse_event <- factor(cd8$adverse_event, levels = c("none","irAE"))

p_umaps <- cd8@meta.data %>% 
    ggplot(aes(x = UMAP_1, 
               y = UMAP_2, 
               color = pm_signature1)) + 
    geom_point(shape = 16) +
    theme_minimal() +
    theme(axis.title = element_blank(),
          axis.text = element_blank(),
          panel.grid = element_blank(),
          text = element_text(size = 15),
          legend.position = "none",
          legend.key.height = unit(0.25,"cm")) +
    labs(color= NULL,
         title = NULL) +
    scale_color_gradient(low = "azure2", high = "blue4") +
  facet_wrap(~adverse_event) +
    guides(color = guide_colorbar(ticks = F)) #removes ticks from colorbar

p_allcells <- cd8@meta.data %>%
    ggplot(aes(x = adverse_event,
               y = pm_signature1)) +
    geom_violin(aes(
        color = adverse_event,
        fill = adverse_event),
        scale = "width",
        adjust = 1,
        trim = F,
        alpha = 0.5,
        draw_quantiles = c(0.25,0.5,0.75),
        lwd = 1) +
    theme_minimal() +
    theme(panel.grid = element_blank(),
          panel.border = element_rect(color = "black", 
                                      fill = NA),
          text = element_text(size = 15),
          legend.position = "right") +
    scale_color_manual(values = c("irAE" = "purple", "none" = "black")) +
    scale_fill_manual(values = c("irAE" = "purple", "none" = "black")) +
    labs(title = "CD8+ T cells blood", 
         y = NULL, 
         x = NULL,
         color = NULL,
         fill = NULL)+ 
  add_pvalue(seurat_integrated@meta.data%>%
                                    t_test(pm_signature1~adverse_event)%>%
                                    add_significance()%>%
                                    add_xy_position(),
                                  label.size=5)


p_patients <- cd8@meta.data %>%
  group_by(adverse_event, orig.ident) %>%
  summarize(pm_signature = mean(pm_signature1)) %>%
    ggplot(aes(x = adverse_event,
               y = pm_signature)) +
    geom_violin(aes(
        color = adverse_event,
        fill = adverse_event),
        scale = "width",
        adjust = 1,
        trim = F,
        alpha = 0.5,
        draw_quantiles = c(0.25,0.5,0.75),
        lwd = 1) +
  geom_point(aes(
        color = adverse_event,
        fill = adverse_event),
        size = 5) +
    theme_minimal() +
    theme(panel.grid = element_blank(),
          panel.border = element_rect(color = "black", 
                                      fill = NA),
          text = element_text(size = 15),
          legend.position = "right") +
    scale_color_manual(values = c("irAE" = "purple", "none" = "black")) +
    scale_fill_manual(values = c("irAE" = "purple", "none" = "black")) +
    labs(title = "CD8+ T cells blood", 
         y = NULL, 
         x = NULL,
         color = NULL,
         fill = NULL)+ 
  add_pvalue(seurat_integrated@meta.data%>%
                 group_by(adverse_event, orig.ident) %>%
  summarize(pm_signature = mean(pm_signature1)) %>% 
    ungroup() %>%
                                    t_test(pm_signature~adverse_event)%>%
                                    add_significance()%>%
                                    add_xy_position() %>%
    filter(p < 0.05),
                                  label.size=5)
    
    grid.arrange(p_umaps, p_allcells, p_patients, widths = c(2,1,1))
```

